{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "faf6f78e-095e-4001-9d36-ec43a6a3c59b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-06-26 16:20:57.350249: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.\n",
      "To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "2023-06-26 16:20:58.313726: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT\n"
     ]
    }
   ],
   "source": [
    "from flair.data import Sentence\n",
    "from flair.nn import Classifier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2e13efe0-f8f3-435b-9c00-3134621272b3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import json\n",
    "import nltk\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "from warnings import filterwarnings\n",
    "import sys\n",
    "sys.path.append(\"/home/aberhe/Projects/SANTAL/ASLR/\")\n",
    "import src.artcileDetail_xml as ad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "73a1fc0b-562f-402b-85be-db72b06d45a0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "file_xml='/data/projects/santal/data/PMC_Articles/oa_comm/xml/all/PMC10000167.xml'#PMC10000014.xml'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e799d777-c530-42e0-a4d2-2a4072bd88c6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "articleDetails=ad.ArticleDetails(file_xml)\n",
    "articleDetails.getDetails()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "98be3bc6-3436-49be-9669-7cb142f74710",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2023-06-26 16:21:10,944 SequenceTagger predicts: Dictionary with 8 tags: <unk>, O, B-Disease, E-Disease, I-Disease, S-Disease, <START>, <STOP>\n",
      "2023-06-26 16:21:11,619 SequenceTagger predicts: Dictionary with 8 tags: <unk>, O, S-Gene, B-Gene, I-Gene, E-Gene, <START>, <STOP>\n",
      "2023-06-26 16:21:12,277 SequenceTagger predicts: Dictionary with 8 tags: <unk>, O, S-Species, B-Species, I-Species, E-Species, <START>, <STOP>\n",
      "2023-06-26 16:21:12,929 SequenceTagger predicts: Dictionary with 8 tags: <unk>, O, S-CellLine, B-CellLine, I-CellLine, E-CellLine, <START>, <STOP>\n",
      "2023-06-26 16:21:13,582 SequenceTagger predicts: Dictionary with 8 tags: <unk>, O, S-Chemical, B-Chemical, I-Chemical, E-Chemical, <START>, <STOP>\n",
      "Warm- and cool-season grasses have different mechanisms for storage of soluble carbohydrates [4], with non-structural carbohydrate (NSC = sugars + starch + fructans) concentrations in cool-season grasses typically greater than that of warm-season grasses [1,5].\n",
      "Span[12:13]: \"carbohydrates\" → Chemical (0.9907)\n",
      "Span[19:20]: \"carbohydrate\" → Chemical (0.9919)\n",
      "Span[23:24]: \"sugars\" → Chemical (0.8969)\n",
      "********************\n",
      "These differences in NSC content are of interest in equine management as current feeding recommendations for horses with existing metabolic dysfunction include limiting dietary NSC concentrations [6,7,8].\n",
      "Span[9:10]: \"equine\" → Species (0.7423)\n",
      "Span[16:17]: \"horses\" → Species (0.9682)\n",
      "Span[19:21]: \"metabolic dysfunction\" → Disease (0.7248)\n",
      "Span[24:25]: \"NSC\" → Chemical (0.661)\n",
      "********************\n",
      "Feeding supplemental concentrate higher in NSC has been shown to lower insulin sensitivity in horses [10,11,12], but over the relatively smaller range of NSC concentrations observed in forages, potential benefits of limiting NSC intake are less clear [13,14,15].\n",
      "Span[5:6]: \"NSC\" → CellLine (0.9263)\n",
      "Span[11:12]: \"insulin\" → Gene (0.9421)\n",
      "Span[14:15]: \"horses\" → Species (0.949)\n",
      "********************\n",
      "However, glycemic and insulinemic responses of horses grazing low-NSC warm-season grasses have not been extensively evaluated.\n",
      "Span[7:8]: \"horses\" → Species (0.9124)\n",
      "********************\n",
      "In addition to the potential metabolic effects, transitioning horses between forage types may also have implications for equine gastrointestinal health.\n",
      "Span[9:10]: \"horses\" → Species (0.9318)\n",
      "********************\n",
      "Diet has been identified as a dominant factor shaping the community structure of the gut microbiota both in humans and across animal species including horses [16,17].\n",
      "Span[18:19]: \"humans\" → Species (0.9154)\n",
      "Span[24:25]: \"horses\" → Species (0.7653)\n",
      "********************\n",
      "However, prior studies on the influence of diet on the equine hindgut microbiome have focused primarily on concentrate vs. concentrate or concentrate vs. forage diets [18,19,20,21].\n",
      "Span[11:12]: \"equine\" → Species (0.7798)\n",
      "********************\n",
      "Recent studies have demonstrated that type of hay (alfalfa vs. grass) and transitions between hay and pasture grass can impact equine cecal or fecal microbial community composition [22,23].\n",
      "Span[9:10]: \"alfalfa\" → Species (0.8291)\n",
      "********************\n",
      "Overall, few studies have evaluated the hindgut microbiome of grazing horses and only one previous study has been conducted in horses grazing cool- vs. warm-season pasture grasses [24].\n",
      "Span[11:12]: \"horses\" → Species (0.6651)\n",
      "Span[21:22]: \"horses\" → Species (0.8042)\n",
      "********************\n",
      "[28] explored relationships between abundance profiles of fecal microbial taxa, feed types (pasture, hay, or hay supplemented with concentrate), and blood analytes including circulating glucose and insulin, finding negative correlations between insulin and over 50 taxa.\n",
      "Span[18:19]: \"hay\" → Species (0.7525)\n",
      "Span[21:22]: \"hay\" → Species (0.5594)\n",
      "Span[32:33]: \"glucose\" → Chemical (0.9902)\n",
      "Span[34:35]: \"insulin\" → Gene (0.9037)\n",
      "Span[40:41]: \"insulin\" → Gene (0.9327)\n",
      "********************\n",
      "Studies conducted in mouse models have conclusively demonstrated that changes in diet influence host metabolism in a microbiome-dependent manner [29,30].\n",
      "Span[3:4]: \"mouse\" → Species (0.9009)\n",
      "********************\n",
      "Therefore, the aims of this study were to characterize shifts in glucose metabolism and the fecal microbiota of horses adapted to different forage types and to explore relationships between forage nutrients, microbial composition, fecal metabolites, and metabolic responses of grazing horses.\n",
      "Span[12:13]: \"glucose\" → Chemical (0.9881)\n",
      "Span[19:20]: \"horses\" → Species (0.81)\n",
      "Span[44:45]: \"horses\" → Species (0.5491)\n",
      "********************\n",
      "In brief, warm-season grass pasture sections contained either Wrangler bermudagrass [BER; Cynodon dactylon (L.) Pers.\n",
      "Span[14:16]: \"Cynodon dactylon\" → Species (0.8459)\n",
      "********************\n",
      "; Johnston Seed Company, Enid, OK, USA] or Quick-N-Big crabgrass [CRB; Digitaria sanguinalis (L.) Scop.\n",
      "Span[13:14]: \"crabgrass\" → Species (0.6588)\n",
      "Span[17:23]: \"Digitaria sanguinalis (L.) Scop\" → Species (0.7807)\n",
      "********************\n",
      "Mixed cool-season grass sections contained Inavale orchardgrass [Dactylis glomerata (L.)], Tower tall fescue (endophyte-free) [Lolium arundinaceum (Schreb.)\n",
      "Span[5:7]: \"Inavale orchardgrass\" → Species (0.8092)\n",
      "Span[8:10]: \"Dactylis glomerata\" → Species (0.9093)\n",
      "Span[21:23]: \"Lolium arundinaceum\" → Species (0.8536)\n",
      "********************\n",
      "], and Argyle Kentucky bluegrass [Poa pratensis (L.)] (DLF Pickseed, Halsey, OR, USA).\n",
      "Span[3:6]: \"Argyle Kentucky bluegrass\" → Species (0.5951)\n",
      "Span[7:9]: \"Poa pratensis\" → Species (0.9417)\n",
      "********************\n",
      "Use of animals in this study was approved by the Rutgers University Institutional Animal Care and Use Committee protocol #PROTO201800013.\n",
      "Span[2:3]: \"animals\" → Species (0.8001)\n",
      "Span[13:14]: \"Animal\" → Species (0.9971)\n",
      "********************\n",
      "Eight adult Standardbred mares (age: 18 ± 0.71 yr; body weight (BW): 537 ± 17 kg; body condition score (BCS): 5–7)) were used in this study, with horses grouped by BW and BCS (Henneke Body Condition Score scale [33].\n",
      "Span[41:42]: \"horses\" → Species (0.9286)\n",
      "********************\n",
      "Horses were then randomly assigned to each system (n = 4 horses system−1).\n",
      "Span[0:1]: \"Horses\" → Species (0.9762)\n",
      "Span[12:13]: \"horses\" → Species (0.7886)\n",
      "********************\n",
      "Prior to the study, oral sugar tests (OST) were administered to screen horses for impaired insulin sensitivity [34].\n",
      "Span[6:7]: \"sugar\" → Chemical (0.8783)\n",
      "Span[15:16]: \"horses\" → Species (0.9812)\n",
      "Span[17:20]: \"impaired insulin sensitivity\" → Disease (0.7141)\n",
      "Span[18:19]: \"insulin\" → Gene (0.9427)\n",
      "********************\n",
      "Insulinemic responses of all horses were found to be normal (peak insulin ≤45 mIU/L) [7].\n",
      "Span[4:5]: \"horses\" → Species (0.9719)\n",
      "Span[12:13]: \"insulin\" → Gene (0.9674)\n",
      "********************\n",
      "Animal care and management have been previously detailed in companion studies [1,24].\n",
      "Span[0:1]: \"Animal\" → Species (0.7984)\n",
      "********************\n",
      "Horse condition measures over the course of the study can be found in Table S2.\n",
      "Span[0:1]: \"Horse\" → Species (0.7013)\n",
      "********************\n",
      "Manual grab fecal samples were collected (0800 h) rectally from horses after 21 d adaptation to the initial hay diet in the spring (HAY-SP), cool-season grass pasture in the spring (CSG-SP), warm-season grass (WSG)—either BER or CRB during the “summer slump” period, and again following a return to cool-season grass in the fall (CSG-FA) and a final cool-season grass hay at the end of the grazing season (HAY-FA).\n",
      "Span[12:13]: \"horses\" → Species (0.8174)\n",
      "********************\n",
      "Upon collection, samples were immediately placed on ice, transported to the laboratory, and then stored at −80 °C.\n",
      "Span[8:9]: \"ice\" → Chemical (0.7841)\n",
      "********************\n",
      "To determine impact of forage type within integrated systems on glycemic and insulinemic responses of grazing horses, OST [34] were conducted following adaptation to CSG-SP, WSG, and HAY-FA.\n",
      "Span[16:17]: \"horses\" → Species (0.6642)\n",
      "********************\n",
      "On the evening prior to each collection period, horses were confined to dry lots at 2000 h with no access to either hay or pasture.\n",
      "Span[9:10]: \"horses\" → Species (0.8947)\n",
      "********************\n",
      "Following the overnight fast, horses were moved into the barn facility at 0800 h. Two consecutive baseline blood samples were collected (15 min apart) by venipuncture at 0800 and an oral dose of Karo Syrup (0.25 mL/kg BW; modified dosing as utilized by Jacob et al.\n",
      "Span[5:6]: \"horses\" → Species (0.9777)\n",
      "********************\n",
      "Whole blood was collected into sodium heparin Vacutainer tubes, which were inverted several times before being placed on ice.\n",
      "Span[5:7]: \"sodium heparin\" → Chemical (0.5496)\n",
      "********************\n",
      "Plasma glucose was analyzed by colorimetric assay (Glucose C-2, Wako Chemicals, Richmond, VA, USA), with the commercial kit adapted for microplate assay following manufacturer instructions.\n",
      "Span[1:2]: \"glucose\" → Chemical (0.992)\n",
      "Span[8:9]: \"Glucose\" → Chemical (0.8099)\n",
      "********************\n",
      "Plasma insulin was evaluated using an enzyme-linked immunoassay (Mercodia Equine Insulin ELISA, Mercodia, Winston-Salem, NC, USA) previously validated in horses [35].\n",
      "Span[1:2]: \"insulin\" → Gene (0.4781)\n",
      "Span[25:26]: \"horses\" → Species (0.9809)\n",
      "********************\n",
      "Inter-assay and intra-assay coefficients of variation for glucose were 4.0% and 2.9%, respectively.\n",
      "Span[7:8]: \"glucose\" → Chemical (0.9843)\n",
      "********************\n",
      "Inter-assay and intra-assay coefficients of variation for insulin were 7.8% and 3.4%, respectively.\n",
      "Span[7:8]: \"insulin\" → Gene (0.9561)\n",
      "********************\n",
      "Short-chain and branched-chain fatty acid (SCFA and BCFA, respectively) concentrations were determined by GC-MS analysis of fecal samples.\n",
      "Span[0:5]: \"Short-chain and branched-chain fatty acid\" → Chemical (0.6859)\n",
      "Span[6:7]: \"SCFA\" → Chemical (0.9396)\n",
      "Span[8:9]: \"BCFA\" → Chemical (0.9533)\n",
      "********************\n",
      "The SCFA analyzed included acetate, propionate, butyrate, and valerate.\n",
      "Span[1:2]: \"SCFA\" → Chemical (0.7808)\n",
      "Span[4:5]: \"acetate\" → Chemical (0.9297)\n",
      "Span[6:7]: \"propionate\" → Chemical (0.9761)\n",
      "Span[8:9]: \"butyrate\" → Chemical (0.9623)\n",
      "Span[11:12]: \"valerate\" → Chemical (0.9409)\n",
      "********************\n",
      "The BCFA analyzed included isobutyrate, isovalerate and isocaproate.\n",
      "Span[1:2]: \"BCFA\" → Chemical (0.9084)\n",
      "Span[4:5]: \"isobutyrate\" → Chemical (0.9633)\n",
      "Span[6:7]: \"isovalerate\" → Chemical (0.9714)\n",
      "Span[8:9]: \"isocaproate\" → Chemical (0.9573)\n",
      "********************\n",
      "In brief, frozen fecal samples were weighed and deposited in bead tubes over dry ice.\n",
      "Span[14:16]: \"dry ice\" → Chemical (0.7996)\n",
      "********************\n",
      "Feces were then resuspended in 1 mL of 0.5% phosphoric acid per 0.1 g of sample and tubes were beaten for 5 min at 22.5 rpm in a cold case.\n",
      "Span[10:12]: \"phosphoric acid\" → Chemical (0.8779)\n",
      "********************\n",
      "Prior to GC-MS analyses, thawed fecal suspensions were re-homogenized and centrifuged (10 min at 17,949× g), with the aqueous phase extracted using diethyl ether in a 1:1 volume to volume ratio.\n",
      "Span[23:24]: \"aqueous\" → Chemical (0.6903)\n",
      "Span[27:29]: \"diethyl ether\" → Chemical (0.9046)\n",
      "********************\n",
      "Before analysis, 2-methyl hexanoic acid (Thermo Fisher Scientific, Dallas, TX, USA) was added to the organic phase extract as an internal standard.\n",
      "Span[3:6]: \"2-methyl hexanoic acid\" → Chemical (0.8794)\n",
      "********************\n",
      "The specifications of the GC-MS system and analyses as well as data acquisition and procedures for SCFA/BCFA quantitation have been previously described by Honarbakhsh et al.\n",
      "Span[16:17]: \"SCFA\" → Chemical (0.7618)\n",
      "Span[18:19]: \"BCFA\" → Chemical (0.9188)\n",
      "********************\n",
      "Animal was considered the experimental unit.\n",
      "Span[0:1]: \"Animal\" → Species (0.7013)\n",
      "********************\n",
      "Glucose and insulin response variables as well as SCFA/BCFA concentrations and fecal pH were analyzed by mixed model ANOVA in R, with grazing system, forage type and their interactions set as fixed factors and horse as the random factor in the initial model.\n",
      "Span[0:1]: \"Glucose\" → Chemical (0.9958)\n",
      "Span[2:3]: \"insulin\" → Gene (0.9242)\n",
      "Span[8:9]: \"SCFA\" → Chemical (0.8729)\n",
      "Span[10:11]: \"BCFA\" → Chemical (0.9582)\n",
      "Span[38:39]: \"horse\" → Species (0.8096)\n",
      "********************\n",
      "Proxies for insulin sensitivity (fasting glucose-to-insulin ratio (FGIR); reciprocal of the square root of insulin (RISQI)) and insulin secretory response (modified insulin-to-glucose ratio (MIRG)) were calculated using baseline (fasting) values as previously described [72,73].\n",
      "Span[2:3]: \"insulin\" → Gene (0.949)\n",
      "Span[6:7]: \"glucose-to-insulin\" → Gene (0.9181)\n",
      "Span[6:7]: \"glucose-to-insulin\" → Chemical (0.7037)\n",
      "Span[18:19]: \"insulin\" → Gene (0.8819)\n",
      "Span[23:24]: \"insulin\" → Gene (0.9078)\n",
      "Span[28:29]: \"insulin-to-glucose\" → Chemical (0.9438)\n",
      "********************\n",
      "The relationship between forage nutrients and SCFA/BCFA fermentation metabolites (as well as pH) and fecal microbial community composition was then explored using random forest regression with nested cross validation to determine if nutrient concentrations and/or metabolite concentrations could be predicted based on bacterial abundance profiles.\n",
      "Span[6:7]: \"SCFA\" → Chemical (0.8605)\n",
      "Span[8:9]: \"BCFA\" → Chemical (0.9266)\n",
      "********************\n",
      "Overall analysis of microbiome, fecal metabolite, and glucose/insulin data did not reveal differences by grazing system.\n",
      "Span[9:10]: \"glucose\" → Chemical (0.9929)\n",
      "Span[11:12]: \"insulin\" → Gene (0.9417)\n",
      "********************\n",
      "In brief, warm-season grass pasture sections contained either Wrangler bermudagrass [BER; Cynodon dactylon (L.) Pers.\n",
      "Span[14:16]: \"Cynodon dactylon\" → Species (0.8459)\n",
      "********************\n",
      "; Johnston Seed Company, Enid, OK, USA] or Quick-N-Big crabgrass [CRB; Digitaria sanguinalis (L.) Scop.\n",
      "Span[13:14]: \"crabgrass\" → Species (0.6588)\n",
      "Span[17:23]: \"Digitaria sanguinalis (L.) Scop\" → Species (0.7807)\n",
      "********************\n",
      "Mixed cool-season grass sections contained Inavale orchardgrass [Dactylis glomerata (L.)], Tower tall fescue (endophyte-free) [Lolium arundinaceum (Schreb.)\n",
      "Span[5:7]: \"Inavale orchardgrass\" → Species (0.8092)\n",
      "Span[8:10]: \"Dactylis glomerata\" → Species (0.9093)\n",
      "Span[21:23]: \"Lolium arundinaceum\" → Species (0.8536)\n",
      "********************\n",
      "], and Argyle Kentucky bluegrass [Poa pratensis (L.)] (DLF Pickseed, Halsey, OR, USA).\n",
      "Span[3:6]: \"Argyle Kentucky bluegrass\" → Species (0.5951)\n",
      "Span[7:9]: \"Poa pratensis\" → Species (0.9417)\n",
      "********************\n",
      "Use of animals in this study was approved by the Rutgers University Institutional Animal Care and Use Committee protocol #PROTO201800013.\n",
      "Span[2:3]: \"animals\" → Species (0.8001)\n",
      "Span[13:14]: \"Animal\" → Species (0.9971)\n",
      "********************\n",
      "Eight adult Standardbred mares (age: 18 ± 0.71 yr; body weight (BW): 537 ± 17 kg; body condition score (BCS): 5–7)) were used in this study, with horses grouped by BW and BCS (Henneke Body Condition Score scale [33].\n",
      "Span[41:42]: \"horses\" → Species (0.9286)\n",
      "********************\n",
      "Horses were then randomly assigned to each system (n = 4 horses system−1).\n",
      "Span[0:1]: \"Horses\" → Species (0.9762)\n",
      "Span[12:13]: \"horses\" → Species (0.7886)\n",
      "********************\n",
      "Prior to the study, oral sugar tests (OST) were administered to screen horses for impaired insulin sensitivity [34].\n",
      "Span[6:7]: \"sugar\" → Chemical (0.8783)\n",
      "Span[15:16]: \"horses\" → Species (0.9812)\n",
      "Span[17:20]: \"impaired insulin sensitivity\" → Disease (0.7141)\n",
      "Span[18:19]: \"insulin\" → Gene (0.9427)\n",
      "********************\n",
      "Insulinemic responses of all horses were found to be normal (peak insulin ≤45 mIU/L) [7].\n",
      "Span[4:5]: \"horses\" → Species (0.9719)\n",
      "Span[12:13]: \"insulin\" → Gene (0.9674)\n",
      "********************\n",
      "Animal care and management have been previously detailed in companion studies [1,24].\n",
      "Span[0:1]: \"Animal\" → Species (0.7984)\n",
      "********************\n",
      "Horse condition measures over the course of the study can be found in Table S2.\n",
      "Span[0:1]: \"Horse\" → Species (0.7013)\n",
      "********************\n",
      "Manual grab fecal samples were collected (0800 h) rectally from horses after 21 d adaptation to the initial hay diet in the spring (HAY-SP), cool-season grass pasture in the spring (CSG-SP), warm-season grass (WSG)—either BER or CRB during the “summer slump” period, and again following a return to cool-season grass in the fall (CSG-FA) and a final cool-season grass hay at the end of the grazing season (HAY-FA).\n",
      "Span[12:13]: \"horses\" → Species (0.8174)\n",
      "********************\n",
      "Upon collection, samples were immediately placed on ice, transported to the laboratory, and then stored at −80 °C.\n",
      "Span[8:9]: \"ice\" → Chemical (0.7841)\n",
      "********************\n",
      "To determine impact of forage type within integrated systems on glycemic and insulinemic responses of grazing horses, OST [34] were conducted following adaptation to CSG-SP, WSG, and HAY-FA.\n",
      "Span[16:17]: \"horses\" → Species (0.6642)\n",
      "********************\n",
      "On the evening prior to each collection period, horses were confined to dry lots at 2000 h with no access to either hay or pasture.\n",
      "Span[9:10]: \"horses\" → Species (0.8947)\n",
      "********************\n",
      "Following the overnight fast, horses were moved into the barn facility at 0800 h. Two consecutive baseline blood samples were collected (15 min apart) by venipuncture at 0800 and an oral dose of Karo Syrup (0.25 mL/kg BW; modified dosing as utilized by Jacob et al.\n",
      "Span[5:6]: \"horses\" → Species (0.9777)\n",
      "********************\n",
      "Whole blood was collected into sodium heparin Vacutainer tubes, which were inverted several times before being placed on ice.\n",
      "Span[5:7]: \"sodium heparin\" → Chemical (0.5496)\n",
      "********************\n",
      "Plasma glucose was analyzed by colorimetric assay (Glucose C-2, Wako Chemicals, Richmond, VA, USA), with the commercial kit adapted for microplate assay following manufacturer instructions.\n",
      "Span[1:2]: \"glucose\" → Chemical (0.992)\n",
      "Span[8:9]: \"Glucose\" → Chemical (0.8099)\n",
      "********************\n",
      "Plasma insulin was evaluated using an enzyme-linked immunoassay (Mercodia Equine Insulin ELISA, Mercodia, Winston-Salem, NC, USA) previously validated in horses [35].\n",
      "Span[1:2]: \"insulin\" → Gene (0.4781)\n",
      "Span[25:26]: \"horses\" → Species (0.9809)\n",
      "********************\n",
      "Inter-assay and intra-assay coefficients of variation for glucose were 4.0% and 2.9%, respectively.\n",
      "Span[7:8]: \"glucose\" → Chemical (0.9843)\n",
      "********************\n",
      "Inter-assay and intra-assay coefficients of variation for insulin were 7.8% and 3.4%, respectively.\n",
      "Span[7:8]: \"insulin\" → Gene (0.9561)\n",
      "********************\n",
      "Short-chain and branched-chain fatty acid (SCFA and BCFA, respectively) concentrations were determined by GC-MS analysis of fecal samples.\n",
      "Span[0:5]: \"Short-chain and branched-chain fatty acid\" → Chemical (0.6859)\n",
      "Span[6:7]: \"SCFA\" → Chemical (0.9396)\n",
      "Span[8:9]: \"BCFA\" → Chemical (0.9533)\n",
      "********************\n",
      "The SCFA analyzed included acetate, propionate, butyrate, and valerate.\n",
      "Span[1:2]: \"SCFA\" → Chemical (0.7808)\n",
      "Span[4:5]: \"acetate\" → Chemical (0.9297)\n",
      "Span[6:7]: \"propionate\" → Chemical (0.9761)\n",
      "Span[8:9]: \"butyrate\" → Chemical (0.9623)\n",
      "Span[11:12]: \"valerate\" → Chemical (0.9409)\n",
      "********************\n",
      "The BCFA analyzed included isobutyrate, isovalerate and isocaproate.\n",
      "Span[1:2]: \"BCFA\" → Chemical (0.9084)\n",
      "Span[4:5]: \"isobutyrate\" → Chemical (0.9633)\n",
      "Span[6:7]: \"isovalerate\" → Chemical (0.9714)\n",
      "Span[8:9]: \"isocaproate\" → Chemical (0.9573)\n",
      "********************\n",
      "In brief, frozen fecal samples were weighed and deposited in bead tubes over dry ice.\n",
      "Span[14:16]: \"dry ice\" → Chemical (0.7996)\n",
      "********************\n",
      "Feces were then resuspended in 1 mL of 0.5% phosphoric acid per 0.1 g of sample and tubes were beaten for 5 min at 22.5 rpm in a cold case.\n",
      "Span[10:12]: \"phosphoric acid\" → Chemical (0.8779)\n",
      "********************\n",
      "Prior to GC-MS analyses, thawed fecal suspensions were re-homogenized and centrifuged (10 min at 17,949× g), with the aqueous phase extracted using diethyl ether in a 1:1 volume to volume ratio.\n",
      "Span[23:24]: \"aqueous\" → Chemical (0.6903)\n",
      "Span[27:29]: \"diethyl ether\" → Chemical (0.9046)\n",
      "********************\n",
      "Before analysis, 2-methyl hexanoic acid (Thermo Fisher Scientific, Dallas, TX, USA) was added to the organic phase extract as an internal standard.\n",
      "Span[3:6]: \"2-methyl hexanoic acid\" → Chemical (0.8794)\n",
      "********************\n",
      "The specifications of the GC-MS system and analyses as well as data acquisition and procedures for SCFA/BCFA quantitation have been previously described by Honarbakhsh et al.\n",
      "Span[16:17]: \"SCFA\" → Chemical (0.7618)\n",
      "Span[18:19]: \"BCFA\" → Chemical (0.9188)\n",
      "********************\n",
      "Animal was considered the experimental unit.\n",
      "Span[0:1]: \"Animal\" → Species (0.7013)\n",
      "********************\n",
      "Glucose and insulin response variables as well as SCFA/BCFA concentrations and fecal pH were analyzed by mixed model ANOVA in R, with grazing system, forage type and their interactions set as fixed factors and horse as the random factor in the initial model.\n",
      "Span[0:1]: \"Glucose\" → Chemical (0.9958)\n",
      "Span[2:3]: \"insulin\" → Gene (0.9242)\n",
      "Span[8:9]: \"SCFA\" → Chemical (0.8729)\n",
      "Span[10:11]: \"BCFA\" → Chemical (0.9582)\n",
      "Span[38:39]: \"horse\" → Species (0.8096)\n",
      "********************\n",
      "Proxies for insulin sensitivity (fasting glucose-to-insulin ratio (FGIR); reciprocal of the square root of insulin (RISQI)) and insulin secretory response (modified insulin-to-glucose ratio (MIRG)) were calculated using baseline (fasting) values as previously described [72,73].\n",
      "Span[2:3]: \"insulin\" → Gene (0.949)\n",
      "Span[6:7]: \"glucose-to-insulin\" → Gene (0.9181)\n",
      "Span[6:7]: \"glucose-to-insulin\" → Chemical (0.7037)\n",
      "Span[18:19]: \"insulin\" → Gene (0.8819)\n",
      "Span[23:24]: \"insulin\" → Gene (0.9078)\n",
      "Span[28:29]: \"insulin-to-glucose\" → Chemical (0.9438)\n",
      "********************\n",
      "The relationship between forage nutrients and SCFA/BCFA fermentation metabolites (as well as pH) and fecal microbial community composition was then explored using random forest regression with nested cross validation to determine if nutrient concentrations and/or metabolite concentrations could be predicted based on bacterial abundance profiles.\n",
      "Span[6:7]: \"SCFA\" → Chemical (0.8605)\n",
      "Span[8:9]: \"BCFA\" → Chemical (0.9266)\n",
      "********************\n",
      "Overall analysis of microbiome, fecal metabolite, and glucose/insulin data did not reveal differences by grazing system.\n",
      "Span[9:10]: \"glucose\" → Chemical (0.9929)\n",
      "Span[11:12]: \"insulin\" → Gene (0.9417)\n",
      "********************\n",
      "Conversely, richness (Observed ASVs) was greater in horses adapted to WSG, CSG-FA, and HAY-FA than CSG-SP (p < 0.02), but did not differ between horses adapted to HAY-SP and CSG-SP.\n",
      "Span[10:11]: \"horses\" → Species (0.5288)\n",
      "Span[32:33]: \"horses\" → Species (0.5761)\n",
      "********************\n",
      "Application of SCNIC identified 333 BCG, with 224 individual ASV remaining ungrouped.\n",
      "Span[10:11]: \"ASV\" → Species (0.6459)\n",
      "********************\n",
      "All retained features were BCG; no ungrouped ASV remained in the reduced feature set.\n",
      "Span[8:9]: \"ASV\" → Chemical (0.6753)\n",
      "********************\n",
      "At the family level, the BCG markers for HAY-SP contained twice as many ASV mapped to the Lachnospiraceae family as any other forage (CSG-SP: 0; WSG: 2; CSG-FA: 3; HAY-FA: 2 ASV).\n",
      "Span[14:15]: \"ASV\" → Species (0.4708)\n",
      "********************\n",
      "At the genus level, ASV assigned to Christensenellaceae R-7 group were present in BCG markers of both hay diets (HAY-SP: 2; HAY-FA:1 ASV) and WSG (2 ASV); ASV assigned to the NK4A214 group of Oscillospiraceae were among members of BCG identified as markers of HAY-SP (2 ASV), WSG (1 ASV), and CSG-FA (1 ASV).\n",
      "Span[5:6]: \"ASV\" → Species (0.5023)\n",
      "Span[37:38]: \"ASV\" → Species (0.6613)\n",
      "********************\n",
      "The BCG markers of both hay diets and CSG-FA each included an ASV mapped to Rikenellaceae RC9 gut group, and BCG markers of HAY-SP and WSG each contained ASV assigned to Fibrobacter and Papillibacter.\n",
      "Span[29:30]: \"ASV\" → Species (0.5787)\n",
      "********************\n",
      "The BCG markers of WSG and CSG-FA included ASV assigned to Bacteriodales RF16 group and the Family XIII AD3011 group of Anaerovoracaceae.\n",
      "Span[8:9]: \"ASV\" → Species (0.5181)\n",
      "********************\n",
      "The BCG markers of CSG-FA and HAY-FA each contained an ASV assigned to Coprostanoligenes group, Lachnospiraceae XPB1014 group and Marvinbryantia, and ASV assigned to Treponema were assigned to BCG markers of WSG and HAY-FA.\n",
      "Span[23:24]: \"ASV\" → Species (0.7979)\n",
      "********************\n",
      "Amplicon sequence variants assigned to Ruminococcus, Pseudobutyvibrio, Anaerovibrio, probable genus 10 of Lachnospiraceae, and the p-251-o5 genus and family of the order Bacteroidales were only present in BCG markers identified for HAY-SP; no other ASV assigned to these taxa were found in BCG markers of other forages.\n",
      "Span[39:40]: \"ASV\" → Species (0.65)\n",
      "********************\n",
      "Amplicon sequence variants mapped to the genera Akkermansia, Mogibacterium, and the Hallii group of Lachnospiraceae as well as UCG-005 metagenome of Oscillospiraceae and Clostridium butyricum were only found in BCG markers of WSG.\n",
      "Span[25:27]: \"Clostridium butyricum\" → Species (0.8221)\n",
      "********************\n",
      "The BCG markers of CSG-FA contained ASV assigned to Alloprevotella, Erysipelatoclostridium, Bacteroidales UCG-001, the WCHB1-41 genus, family and class within the class Kiritimatiellae, and Denitrobacterium detoxificans; these taxa were not identified in BCG markers of other forages.\n",
      "Span[6:7]: \"ASV\" → Species (0.5132)\n",
      "Span[29:31]: \"Denitrobacterium detoxificans\" → Species (0.6463)\n",
      "********************\n",
      "Taxa specific to only BCG markers of HAY-FA included ASV from the Incertae Sedis genus of Ethanoligenenaceae as well as Streptococcus.\n",
      "Span[9:10]: \"ASV\" → Species (0.5935)\n",
      "Span[20:21]: \"Streptococcus\" → Species (0.82)\n",
      "********************\n",
      "Fecal pH was greater in horses adapted to WSG (7.56 ± 0.18) and HAY-FA (7.57 ± 0.18) than in HAY-SP (6.73 ± 0.18), CSG-SP (6.58 ± 0.18), or CSG-FA (6.53 ± 0.18; p ≤ 0.02; Figure 3).\n",
      "Span[5:6]: \"horses\" → Species (0.8445)\n",
      "********************\n",
      "Fecal concentrations of short-chain fatty acids (SCFA) and branched-chain fatty acids (BCFA) also differed by forage (mixed model ANOVA with Tukey’s post hoc adjustment; p < 0.001; Table 6).\n",
      "Span[3:6]: \"short-chain fatty acids\" → Chemical (0.6761)\n",
      "Span[7:8]: \"SCFA\" → Chemical (0.9869)\n",
      "Span[10:13]: \"branched-chain fatty acids\" → Chemical (0.7418)\n",
      "Span[14:15]: \"BCFA\" → Chemical (0.9694)\n",
      "********************\n",
      "Unlike fecal pH, differences in SCFA and BCFA were primarily between pasture forages and the hay diets, with horses adapted to WSG often intermediate (numerically) between cool-season pasture and hay diets.\n",
      "Span[6:7]: \"SCFA\" → Chemical (0.9544)\n",
      "Span[8:9]: \"BCFA\" → Chemical (0.957)\n",
      "Span[20:21]: \"horses\" → Species (0.8734)\n",
      "********************\n",
      "Total BCFA were greater in CSG-SP, WSG, and CSG-FA than for either HAY-SP or HAY-FA (p < 0.05).\n",
      "Span[1:2]: \"BCFA\" → Chemical (0.9507)\n",
      "********************\n",
      "Total SCFA were greater in CSG-SP and CSG-FA than HAY-SP or HAY-FA; WSG did not differ from CSG-SP and CSG-FA or HAY-SP, but was greater in comparison to HAY-FA (p < 0.002).\n",
      "Span[1:2]: \"SCFA\" → Chemical (0.8532)\n",
      "********************\n",
      "Similarly, acetate was greater in CSG-FA than HAY-SP or HAY-FA, while WSG only differed from HAY-FA (p < 0.02).\n",
      "Span[2:3]: \"acetate\" → Chemical (0.904)\n",
      "********************\n",
      "Acetate concentrations for CSG-FA also differed with HAY-FA (p = 0.0003), but there was only a trend for a difference between CSG-SP and HAY-SP (p = 0.09).\n",
      "Span[0:1]: \"Acetate\" → Chemical (0.9293)\n",
      "********************\n",
      "Fecal butyrate was lower for HAY-FA than when horses were adapted to any of the pasture forages (p < 0.03), but there was no difference between HAY-SP and any other forage.\n",
      "Span[1:2]: \"butyrate\" → Chemical (0.9568)\n",
      "Span[8:9]: \"horses\" → Species (0.7168)\n",
      "********************\n",
      "Fecal propionate was lowest in horses adapted to HAY-FA (p < 0.002), but propionate concentrations were also lower for HAY-SP than either CSG-SP or CSG-FA (p < 0.03).\n",
      "Span[1:2]: \"propionate\" → Chemical (0.9449)\n",
      "Span[5:6]: \"horses\" → Species (0.9339)\n",
      "Span[16:17]: \"propionate\" → Chemical (0.9499)\n",
      "********************\n",
      "Propionate did not differ between horses adapted to WSG and all other forages.\n",
      "Span[0:1]: \"Propionate\" → Chemical (0.919)\n",
      "Span[5:6]: \"horses\" → Species (0.8991)\n",
      "********************\n",
      "Fecal valerate was lower for horses adapted to HAY-FA than CSG-SP or CSG-FA (p < 0.03) but did not differ from WSG or HAY-SP.\n",
      "Span[1:2]: \"valerate\" → Chemical (0.8694)\n",
      "Span[5:6]: \"horses\" → Species (0.8234)\n",
      "********************\n",
      "Horses adapted to HAY-SP also had fecal valerate concentrations lower than CSG-SP or CSG-FA (p ≤ 0.0002).\n",
      "Span[0:1]: \"Horses\" → Species (0.9317)\n",
      "Span[7:8]: \"valerate\" → Chemical (0.9052)\n",
      "********************\n",
      "Isobutyrate was greater for all pasture forages in comparison to both HAY-SP and HAY-FA (p < 0.02).\n",
      "Span[0:1]: \"Isobutyrate\" → Chemical (0.9504)\n",
      "********************\n",
      "Isovalerate was greater in CSG-SP and CSG-FA vs. HAY-SP and HAY-FA (p ≤ 0.002), while concentrations in horses adapted to WSG were once again intermediate.\n",
      "Span[0:1]: \"Isovalerate\" → Chemical (0.9601)\n",
      "Span[20:21]: \"horses\" → Species (0.9257)\n",
      "********************\n",
      "Isocaproate was detected in all fecal samples, but concentrations were negligible and below the limit of quantification (<1.0 ug g feces−1).\n",
      "Span[0:1]: \"Isocaproate\" → Chemical (0.97)\n",
      "********************\n",
      "Relatively weak predictive accuracy was found for isovalerate, valerate, hexanoate, and heptanoate (model R2 and p-values are shown in Table 7).\n",
      "Span[7:8]: \"isovalerate\" → Chemical (0.9498)\n",
      "Span[9:10]: \"valerate\" → Chemical (0.9548)\n",
      "Span[11:12]: \"hexanoate\" → Chemical (0.9617)\n",
      "Span[14:15]: \"heptanoate\" → Chemical (0.9283)\n",
      "********************\n",
      "Major SCFA including acetate, butyrate, and propionate could not be predicted based on BCG abundance profiles.\n",
      "Span[1:2]: \"SCFA\" → Chemical (0.9484)\n",
      "Span[3:4]: \"acetate\" → Chemical (0.961)\n",
      "Span[5:6]: \"butyrate\" → Chemical (0.9745)\n",
      "Span[8:9]: \"propionate\" → Chemical (0.9387)\n",
      "********************\n",
      "However, when random forest regressors were applied to only the top 25 BCG identified as most predictive of forage type, model accuracies improved for all fecal variables with the exception of valerate (Table 7).\n",
      "Span[33:34]: \"valerate\" → Chemical (0.9062)\n",
      "********************\n",
      "There were positive correlations between BCG_300 and total SCFA and total BCFA as well as with acetate, butyrate, propionate, valerate, isobutyrate, and isovalerate, while this BCG was negatively correlated with hexanoate, heptanoate, and pH.\n",
      "Span[10:11]: \"SCFA\" → Chemical (0.9694)\n",
      "Span[13:14]: \"BCFA\" → Chemical (0.987)\n",
      "Span[18:19]: \"acetate\" → Chemical (0.9385)\n",
      "Span[20:21]: \"butyrate\" → Chemical (0.9803)\n",
      "Span[22:23]: \"propionate\" → Chemical (0.9761)\n",
      "Span[24:25]: \"valerate\" → Chemical (0.9773)\n",
      "Span[26:27]: \"isobutyrate\" → Chemical (0.9627)\n",
      "Span[29:30]: \"isovalerate\" → Chemical (0.953)\n",
      "Span[38:39]: \"hexanoate\" → Chemical (0.9347)\n",
      "Span[40:41]: \"heptanoate\" → Chemical (0.9496)\n",
      "********************\n",
      "The ASV members of this BCG were assigned to the genera Lachnospiraceae XPB1014 and Streptococcus.\n",
      "Span[14:15]: \"Streptococcus\" → Species (0.7771)\n",
      "********************\n",
      "There were also positive correlations between both BCG_9 and BCG_259 and total SCFA and BCFA in addition to acetate, butyrate, propionate, valerate, isobutyrate, and isovalerate; these BCG were negatively correlated with heptoanate.\n",
      "Span[16:17]: \"SCFA\" → Chemical (0.9858)\n",
      "Span[18:19]: \"BCFA\" → Chemical (0.9927)\n",
      "Span[22:23]: \"acetate\" → Chemical (0.9331)\n",
      "Span[24:25]: \"butyrate\" → Chemical (0.9812)\n",
      "Span[26:27]: \"propionate\" → Chemical (0.9786)\n",
      "Span[28:29]: \"valerate\" → Chemical (0.9805)\n",
      "Span[30:31]: \"isobutyrate\" → Chemical (0.9661)\n",
      "Span[33:34]: \"isovalerate\" → Chemical (0.9603)\n",
      "Span[41:42]: \"heptoanate\" → Chemical (0.9253)\n",
      "********************\n",
      "These BCG included multiple ASV assigned to Lachnospiraceae as well as ASV within the genera Christensenellaceae R-7 group, Rikenellaceae RC9 gut group, Papillibacter, and the NK4214 group of Oscillospiraceae.\n",
      "Span[4:5]: \"ASV\" → Species (0.6397)\n",
      "Span[11:12]: \"ASV\" → Species (0.6306)\n",
      "********************\n",
      "Positive correlations were also found between BCG_173 and BCG_201 and total SCFA and BCFA, acetate, propionate, valerate, isobutyrate, and isovalerate, with negative correlations between these BCG and both hexoanate and heptoanate.\n",
      "Span[15:16]: \"SCFA\" → Chemical (0.9843)\n",
      "Span[17:18]: \"BCFA\" → Chemical (0.997)\n",
      "Span[19:20]: \"acetate\" → Chemical (0.9545)\n",
      "Span[21:22]: \"propionate\" → Chemical (0.9774)\n",
      "Span[23:24]: \"valerate\" → Chemical (0.979)\n",
      "Span[25:26]: \"isobutyrate\" → Chemical (0.9646)\n",
      "Span[28:29]: \"isovalerate\" → Chemical (0.9594)\n",
      "Span[35:36]: \"BCG\" → Chemical (0.7793)\n",
      "Span[38:39]: \"hexoanate\" → Chemical (0.9363)\n",
      "Span[40:41]: \"heptoanate\" → Chemical (0.9544)\n",
      "********************\n",
      "These BCG included ASV assigned to Treponema, Rikenellaceae RC9 gut group, the NKA214 group of Oscillospiraceae, and the p-251-o5 and F082 genera and families of Bacteroidales.\n",
      "Span[3:4]: \"ASV\" → Species (0.7001)\n",
      "********************\n",
      "Total SCFA and BCFA as well as acetate, butyrate, propionate, isobutyrate, and isovalerate were positively correlated with BCG_113, while hexanoate was negatively correlated.\n",
      "Span[1:2]: \"SCFA\" → Chemical (0.9723)\n",
      "Span[3:4]: \"BCFA\" → Chemical (0.9781)\n",
      "Span[7:8]: \"acetate\" → Chemical (0.9366)\n",
      "Span[9:10]: \"butyrate\" → Chemical (0.9758)\n",
      "Span[11:12]: \"propionate\" → Chemical (0.971)\n",
      "Span[13:14]: \"isobutyrate\" → Chemical (0.9627)\n",
      "Span[16:17]: \"isovalerate\" → Chemical (0.9407)\n",
      "Span[26:27]: \"hexanoate\" → Chemical (0.9399)\n",
      "********************\n",
      "This BCG contained ASV assigned to taxa including Synergistaceae, Bacteroidales RF16 group, Papillibacter, Christensenellaceae R-7 Group, Mogibacterium, and Clostridium butyricum.\n",
      "Span[3:4]: \"ASV\" → Species (0.5622)\n",
      "Span[23:25]: \"Clostridium butyricum\" → Species (0.7977)\n",
      "********************\n",
      "Seven BCG were positively correlated with total BCFA and some combination of propionate, valerate, isobutyrate, and isovalerate, while negative correlations were found between these BCG and hexanoate and/or heptanoate.\n",
      "Span[7:8]: \"BCFA\" → Chemical (0.9777)\n",
      "Span[12:13]: \"propionate\" → Chemical (0.9466)\n",
      "Span[14:15]: \"valerate\" → Chemical (0.9701)\n",
      "Span[16:17]: \"isobutyrate\" → Chemical (0.9546)\n",
      "Span[19:20]: \"isovalerate\" → Chemical (0.951)\n",
      "Span[28:29]: \"BCG\" → Chemical (0.8889)\n",
      "Span[30:31]: \"hexanoate\" → Chemical (0.8779)\n",
      "Span[34:35]: \"heptanoate\" → Chemical (0.9165)\n",
      "********************\n",
      "These BCG included ASV members assigned to the family level for Anaerovoracaceae, Oscillospiraceae, and Lachnospiraceae.\n",
      "Span[3:4]: \"ASV\" → Species (0.561)\n",
      "********************\n",
      "Other ASV within these BCG were assigned to Lactobacillus equigenerosi and the genera Akkermansia, Fibrobacter, Treponema, Sphaerocheata, Phasolarctobacterium, Christensenellaceae R-7 group, Lachnoclostridium, Cellulosilyticum, Marvinbryantia, Coprostanoligenes group, Lachnospiraceae XPB1014, Lachnospiraceae UCG-009, Erysipelatoclostridium, Clostridium sensu stricto 1, and Bacteroidales BS11 gut group as well as Family XIII AD3011 group of Anaerovoracaceae, the UCG-004 genus within Erysipelatoclostridiaceae, the UCG-002 and UCG-005 genuses of Oscillospiraceae, and the UCG-010 genus and family of Oscillospirales.\n",
      "Span[8:10]: \"Lactobacillus equigenerosi\" → Species (0.6602)\n",
      "********************\n",
      "This BCG included ASV assigned to Prevotella, Sarcina, and Clostridium sensu stricto 1.\n",
      "Span[3:4]: \"ASV\" → Species (0.6284)\n",
      "********************\n",
      "Application of random forest regressors demonstrated that forage nutrient concentrations could be predicted based on bacterial community composition (BCG composition of the full microbial community), including water-soluble carbohydrate (WSC) and NSC at a predictive accuracy > R2 = 0.50 and crude protein (CP) with a predictive accuracy of R2 = 0.41 (p < 0.0001; Table 7).\n",
      "Span[30:31]: \"carbohydrate\" → Chemical (0.9793)\n",
      "Span[35:36]: \"NSC\" → CellLine (0.8274)\n",
      "********************\n",
      "Similar to results for fecal metabolites, when random forest regressors were applied only to the top 25 BCG identified as most predictive of forage type, model accuracies improved (Table 7).\n",
      "Span[5:6]: \"metabolites\" → Chemical (0.7674)\n",
      "********************\n",
      "There were distinct ASV from the taxa Christensenellaceae R-7 group and NK4A214 group of Oscillospiraceae that were members of separate BCG which were positively and negatively responding to these nutrients (i.e., these taxonomic classifications were found across all positive and negative responder groups).\n",
      "Span[3:4]: \"ASV\" → Species (0.5328)\n",
      "********************\n",
      "The BCG negatively correlated with ADF/NDF, and thus positively correlated with CP/DE included ASV members assigned to Clostridium butyricum and genera Papillibacter, Lachnoclostridium, Cellulosilyticum, Fibrobacter, Treponema, Catenisphaera, Bacteroidales RF16 group, Synergistaceae, Mogibacterium, Rikenellaceae RC9 gut group, the Hallii group of Lachnospiraceae, the p-251-o5 genus and family of Oscillospiraceae, and the F082 genus and family of Bacteroidales, as well as multiple ASV assigned only to the family level for Lachnospiraceae.\n",
      "Span[18:19]: \"ASV\" → Species (0.4645)\n",
      "Span[22:24]: \"Clostridium butyricum\" → Species (0.847)\n",
      "Span[78:79]: \"ASV\" → Species (0.4972)\n",
      "********************\n",
      "The BCG positively correlated with ADF/NDF, and negatively correlated with CP/DE included ASV within Bacteroidales BS11 gut group and Clostridium sensu stricto 1.\n",
      "Span[24:27]: \"Clostridium sensu stricto\" → Species (0.6194)\n",
      "********************\n",
      "Forage ADF and NDF were also negatively correlated with BCG_300 (Streptococcus and Lachnospiraceae XPB1014), but this BCG was not correlated with either CP or DE.\n",
      "Span[13:14]: \"Streptococcus\" → Species (0.8052)\n",
      "********************\n",
      "The ASV members of BCG_35 were assigned to taxa including Akkermansia, Fibrobacter, Treponema, Christensenellaceae R-7 group, and Family XIII AD3011 within Anaerovoracaceae.\n",
      "Span[18:21]: \"Christensenellaceae R-7 group\" → Species (0.7608)\n",
      "Span[23:26]: \"Family XIII AD3011\" → Species (0.734)\n",
      "********************\n",
      "Forage NSC and WSC were correlated with 13 and 12 BCG, respectively (rs ≥ |0.30|; p ≤ 0.05).\n",
      "Span[1:2]: \"NSC\" → CellLine (0.7083)\n",
      "********************\n",
      "In addition to the above-mentioned taxa from ASV within BCG_35, ASV within BCG negatively correlated with both NSC and WSC were assigned to Lactobacillus equigenerosi and genera including Lachnoclostridium, Cellulosilyticum, Catenisphaera, Marvinbryantia, Erysipelatoclostridium, Mogibacterium, Lachnospiraceae XPB1014 group, Lachnospiraceae UCG-009, Coprostanoligenes group, UCG-005 metagenome within Oscillospiraceae, the Hallii group within Lachnospiraceae, the NK4A214 group of Oscillospiraceae, the Incertae Sedis genus of Ethanoligenenaceae, the WCHB1-41 genus, family, and order within Kiritimatiellae, as well as Denitrobacterium detoxificans.\n",
      "Span[91:93]: \"Denitrobacterium detoxificans\" → Species (0.6941)\n",
      "********************\n",
      "Forage NSC and WSC were positively correlated with BCG_94 (Clostridium sensu stricto 1, Prevotella, and Sarcina).\n",
      "Span[12:15]: \"Clostridium sensu stricto\" → Species (0.6416)\n",
      "********************\n",
      "Forage ESC was correlated with 9 of the same BCG as NSC and WSC, but also had a negative correlation with BCG_173, which contained multiple ASV assigned to Lachnospiraceae, probable genus 10 within Lachnospiraceae, Pseudobutyvibrio as well as to the p-251 genus and family of Bacteroidales.\n",
      "Span[22:25]: \"BCG_173\" → Species (0.6357)\n",
      "********************\n",
      "Acetate, butyrate, propionate, valerate, isobutyrate, isovalerate and total SCFA and BCFA were positively correlated with DE (Spearman correlation; rs ≥ 0.62) and CP (rs ≥ 0.41), but were negatively correlated with NDF (rs ≤ −0.48) and ADF (rs ≤ −0.57; p ≤ 0.008).\n",
      "Span[0:1]: \"Acetate\" → Chemical (0.9634)\n",
      "Span[2:3]: \"butyrate\" → Chemical (0.9781)\n",
      "Span[4:5]: \"propionate\" → Chemical (0.9698)\n",
      "Span[6:7]: \"valerate\" → Chemical (0.9681)\n",
      "Span[8:9]: \"isobutyrate\" → Chemical (0.9688)\n",
      "Span[10:11]: \"isovalerate\" → Chemical (0.9393)\n",
      "Span[13:14]: \"SCFA\" → Chemical (0.9571)\n",
      "Span[15:16]: \"BCFA\" → Chemical (0.9574)\n",
      "********************\n",
      "Total BCFA, isobutyrate, and isovalerate were negatively correlated with NSC, WSC, and ESC (rs ≤ −0.37; p ≤ 0.02), with a weaker negative correlation between valerate and NSC and WSC (rs ≤ −0.34; p ≤ 0.03).\n",
      "Span[1:2]: \"BCFA\" → Chemical (0.9924)\n",
      "Span[3:4]: \"isobutyrate\" → Chemical (0.9497)\n",
      "Span[6:7]: \"isovalerate\" → Chemical (0.9405)\n",
      "Span[34:35]: \"valerate\" → Chemical (0.9389)\n",
      "********************\n",
      "Hexanoate and heptanoate were negatively correlated with DE, CP, and starch (rs ≤ −0.39; p ≤ 0.01), but were positively correlated with NDF, ADF, NSC, WSC, and ESC (rs ≥ 0.37; p ≤ 0.02).\n",
      "Span[0:1]: \"Hexanoate\" → Chemical (0.9446)\n",
      "Span[2:3]: \"heptanoate\" → Chemical (0.916)\n",
      "********************\n",
      "Plasma glucose responses to OST administration differed by forage (mixed model ANOVA with Tukey’s post hoc adjustment; p ≤ 0.01).\n",
      "Span[1:2]: \"glucose\" → Chemical (0.9944)\n",
      "********************\n",
      "The AUC for glucose was lowest when horses were adapted to WSG (42.4 ± 6.8 mg/dL*h) vs. CSG-SP (70.0 ± 6.8 mg/dL*h) and HAY-FA (76.3 ± 6.8 mg/dL*h; p ≤ 0.03; Figure 7a).\n",
      "Span[3:4]: \"glucose\" → Chemical (0.9884)\n",
      "Span[7:8]: \"horses\" → Species (0.968)\n",
      "********************\n",
      "Peak plasma glucose was also lower for WSG (110 ± 3 mg/dL) in comparison to HAY-FA (123 ± 3 mg/dL; p = 0.01), and there was a trend for lower peak plasma glucose for WSG than for CSG-SP (120 ± 3 mg/dL; p = 0.06; Figure 7b).\n",
      "Span[2:3]: \"glucose\" → Chemical (0.9947)\n",
      "Span[42:43]: \"glucose\" → Chemical (0.9962)\n",
      "********************\n",
      "Forage did not, however, impact OST insulin responses.\n",
      "Span[8:9]: \"insulin\" → Gene (0.8972)\n",
      "********************\n",
      "Neither AUC (CSG-SP: 36.1; WSG: 29.1; HAY-FA: 32.7 ± 6.4 mIU/L*h) or peak plasma insulin (CSG-SP: 28.8; WSG: 25.5; HAY-FA: 26.6 ± 3.3 mIU/L) varied by forage (Figure S2).\n",
      "Span[25:26]: \"insulin\" → Gene (0.5557)\n",
      "********************\n",
      "Fasting plasma insulin (CSG-SP: 3.69; WSG: 5.04; HAY-FA: 5.20 ± 0.64 mIU/L) and glucose (CSG-SP: 81.4; WSG: 78.7; HAY-FA: 81.3 ± 1.4 mg/dL) also did not differ by forage (Figure S3).\n",
      "Span[2:3]: \"insulin\" → Gene (0.5653)\n",
      "Span[22:23]: \"glucose\" → Chemical (0.9896)\n",
      "********************\n",
      "There were trends for differences by forage in proxies for insulin sensitivity including the fasting glucose-to-insulin ratio (FGIR) and reciprocal of the square root of insulin (RISQI; mixed model ANOVA with Tukey’s post hoc adjustment; p ≤ 0.10; Figure S4).\n",
      "Span[10:12]: \"insulin sensitivity\" → Disease (0.6053)\n",
      "Span[10:11]: \"insulin\" → Gene (0.9635)\n",
      "Span[15:16]: \"glucose-to-insulin\" → Gene (0.9613)\n",
      "Span[27:28]: \"insulin\" → Gene (0.9114)\n",
      "********************\n",
      "However, the modified insulin-to-glucose ratio (MIRG), a proxy for the insulin secretory response, did not vary by forage (Figure S4).\n",
      "Span[4:5]: \"insulin-to-glucose\" → Chemical (0.9082)\n",
      "Span[14:15]: \"insulin\" → Gene (0.8769)\n",
      "********************\n",
      "Glucose and insulin dynamics were only assessed after three of the forages (CSG-SP, WSG, and HAY-FA), and this smaller dataset precluded the use of random forest regression modelling for these variables.\n",
      "Span[0:1]: \"Glucose\" → Chemical (0.9955)\n",
      "Span[2:3]: \"insulin\" → Gene (0.9326)\n",
      "********************\n",
      "As the primary metabolic differences in grazing horse metabolism were AUC and peak plasma glucose in response to OST administration, correlation analysis was conducted to explore relationships between these metabolic variables and BCG.\n",
      "Span[7:8]: \"horse\" → Species (0.7658)\n",
      "Span[14:15]: \"glucose\" → Chemical (0.9949)\n",
      "********************\n",
      "Only one BCG (of the 25 BCG in the reduced feature set) was correlated with AUC (rs = −0.48; p = 0.04), with members of BCG_124 including ASV within Papillibacter, Christensenellaceae R-7 group, and Synergistaceae.\n",
      "Span[32:35]: \"BCG_124\" → Species (0.6048)\n",
      "Span[36:37]: \"ASV\" → Species (0.7066)\n",
      "********************\n",
      "Four additional BCG were correlated with peak plasma glucose.\n",
      "Span[8:9]: \"glucose\" → Chemical (0.9915)\n",
      "********************\n",
      "Positive correlations were found between peak glucose and BCG_51 and BCG_305 (rs ≥ 0.43; p ≤ 0.04).\n",
      "Span[6:7]: \"glucose\" → Chemical (0.9896)\n",
      "********************\n",
      "These BCG included ASV mapped to genera including Lachnospiraceae XPB1014, Lachnospiraceae UCG-009, Erysipelatoclostridium, Catenisphaera, and Fibrobacter as well to the family level for Anaerovoracaceae.\n",
      "Span[3:4]: \"ASV\" → Species (0.5558)\n",
      "********************\n",
      "Conversely, BCG_113 (Clostridium butyricum, Papillibacter, Bacteroidales RF16 group) and BCG_259 were negatively correlated with peak glucose (rs = −0.44; p = 0.03).\n",
      "Span[6:8]: \"Clostridium butyricum\" → Species (0.7287)\n",
      "Span[24:25]: \"glucose\" → Chemical (0.9726)\n",
      "********************\n",
      "Members of BCG_259 included ASV assigned at the family level to Lachnospiraceae in addition to the Hallii group within Lachnospiraceae.\n",
      "Span[6:7]: \"ASV\" → Species (0.5467)\n",
      "********************\n",
      "While relationships between AUC and peak plasma glucose and forage nutrients were identified through correlation analysis (Spearman correlation; rs ≥ |0.41|; p ≤ 0.04; Figure 7c), these metabolic variables were not correlated with any fecal variables including SCFA, BCFA, and pH.\n",
      "Span[7:8]: \"glucose\" → Chemical (0.9939)\n",
      "Span[45:46]: \"SCFA\" → Chemical (0.9549)\n",
      "Span[47:48]: \"BCFA\" → Chemical (0.9878)\n",
      "********************\n",
      "Forage NSC, WSC, and ESC were positively correlated with AUC (rs ≥ 0.53; p ≤ 0.007) and peak glucose (rs ≥ 0.41; p ≤ 0.04).\n",
      "Span[1:2]: \"NSC\" → CellLine (0.6574)\n",
      "Span[23:24]: \"glucose\" → Chemical (0.9854)\n",
      "********************\n",
      "There was also a negative correlation between CP and glucose responses to OST administration (rs ≤ −0.50; p ≤ 0.01).\n",
      "Span[9:10]: \"glucose\" → Chemical (0.9886)\n",
      "********************\n",
      "Forage DE, NDF, ADF, and starch, however, were not correlated with AUC or peak glucose.\n",
      "Span[19:20]: \"glucose\" → Chemical (0.9774)\n",
      "********************\n",
      "Conversely, richness (Observed ASVs) was greater in horses adapted to WSG, CSG-FA, and HAY-FA than CSG-SP (p < 0.02), but did not differ between horses adapted to HAY-SP and CSG-SP.\n",
      "Span[10:11]: \"horses\" → Species (0.5288)\n",
      "Span[32:33]: \"horses\" → Species (0.5761)\n",
      "********************\n",
      "Application of SCNIC identified 333 BCG, with 224 individual ASV remaining ungrouped.\n",
      "Span[10:11]: \"ASV\" → Species (0.6459)\n",
      "********************\n",
      "All retained features were BCG; no ungrouped ASV remained in the reduced feature set.\n",
      "Span[8:9]: \"ASV\" → Chemical (0.6753)\n",
      "********************\n",
      "At the family level, the BCG markers for HAY-SP contained twice as many ASV mapped to the Lachnospiraceae family as any other forage (CSG-SP: 0; WSG: 2; CSG-FA: 3; HAY-FA: 2 ASV).\n",
      "Span[14:15]: \"ASV\" → Species (0.4708)\n",
      "********************\n",
      "At the genus level, ASV assigned to Christensenellaceae R-7 group were present in BCG markers of both hay diets (HAY-SP: 2; HAY-FA:1 ASV) and WSG (2 ASV); ASV assigned to the NK4A214 group of Oscillospiraceae were among members of BCG identified as markers of HAY-SP (2 ASV), WSG (1 ASV), and CSG-FA (1 ASV).\n",
      "Span[5:6]: \"ASV\" → Species (0.5023)\n",
      "Span[37:38]: \"ASV\" → Species (0.6613)\n",
      "********************\n",
      "The BCG markers of both hay diets and CSG-FA each included an ASV mapped to Rikenellaceae RC9 gut group, and BCG markers of HAY-SP and WSG each contained ASV assigned to Fibrobacter and Papillibacter.\n",
      "Span[29:30]: \"ASV\" → Species (0.5787)\n",
      "********************\n",
      "The BCG markers of WSG and CSG-FA included ASV assigned to Bacteriodales RF16 group and the Family XIII AD3011 group of Anaerovoracaceae.\n",
      "Span[8:9]: \"ASV\" → Species (0.5181)\n",
      "********************\n",
      "The BCG markers of CSG-FA and HAY-FA each contained an ASV assigned to Coprostanoligenes group, Lachnospiraceae XPB1014 group and Marvinbryantia, and ASV assigned to Treponema were assigned to BCG markers of WSG and HAY-FA.\n",
      "Span[23:24]: \"ASV\" → Species (0.7979)\n",
      "********************\n",
      "Amplicon sequence variants assigned to Ruminococcus, Pseudobutyvibrio, Anaerovibrio, probable genus 10 of Lachnospiraceae, and the p-251-o5 genus and family of the order Bacteroidales were only present in BCG markers identified for HAY-SP; no other ASV assigned to these taxa were found in BCG markers of other forages.\n",
      "Span[39:40]: \"ASV\" → Species (0.65)\n",
      "********************\n",
      "Amplicon sequence variants mapped to the genera Akkermansia, Mogibacterium, and the Hallii group of Lachnospiraceae as well as UCG-005 metagenome of Oscillospiraceae and Clostridium butyricum were only found in BCG markers of WSG.\n",
      "Span[25:27]: \"Clostridium butyricum\" → Species (0.8221)\n",
      "********************\n",
      "The BCG markers of CSG-FA contained ASV assigned to Alloprevotella, Erysipelatoclostridium, Bacteroidales UCG-001, the WCHB1-41 genus, family and class within the class Kiritimatiellae, and Denitrobacterium detoxificans; these taxa were not identified in BCG markers of other forages.\n",
      "Span[6:7]: \"ASV\" → Species (0.5132)\n",
      "Span[29:31]: \"Denitrobacterium detoxificans\" → Species (0.6463)\n",
      "********************\n",
      "Taxa specific to only BCG markers of HAY-FA included ASV from the Incertae Sedis genus of Ethanoligenenaceae as well as Streptococcus.\n",
      "Span[9:10]: \"ASV\" → Species (0.5935)\n",
      "Span[20:21]: \"Streptococcus\" → Species (0.82)\n",
      "********************\n",
      "Fecal pH was greater in horses adapted to WSG (7.56 ± 0.18) and HAY-FA (7.57 ± 0.18) than in HAY-SP (6.73 ± 0.18), CSG-SP (6.58 ± 0.18), or CSG-FA (6.53 ± 0.18; p ≤ 0.02; Figure 3).\n",
      "Span[5:6]: \"horses\" → Species (0.8445)\n",
      "********************\n",
      "Fecal concentrations of short-chain fatty acids (SCFA) and branched-chain fatty acids (BCFA) also differed by forage (mixed model ANOVA with Tukey’s post hoc adjustment; p < 0.001; Table 6).\n",
      "Span[3:6]: \"short-chain fatty acids\" → Chemical (0.6761)\n",
      "Span[7:8]: \"SCFA\" → Chemical (0.9869)\n",
      "Span[10:13]: \"branched-chain fatty acids\" → Chemical (0.7418)\n",
      "Span[14:15]: \"BCFA\" → Chemical (0.9694)\n",
      "********************\n",
      "Unlike fecal pH, differences in SCFA and BCFA were primarily between pasture forages and the hay diets, with horses adapted to WSG often intermediate (numerically) between cool-season pasture and hay diets.\n",
      "Span[6:7]: \"SCFA\" → Chemical (0.9544)\n",
      "Span[8:9]: \"BCFA\" → Chemical (0.957)\n",
      "Span[20:21]: \"horses\" → Species (0.8734)\n",
      "********************\n",
      "Total BCFA were greater in CSG-SP, WSG, and CSG-FA than for either HAY-SP or HAY-FA (p < 0.05).\n",
      "Span[1:2]: \"BCFA\" → Chemical (0.9507)\n",
      "********************\n",
      "Total SCFA were greater in CSG-SP and CSG-FA than HAY-SP or HAY-FA; WSG did not differ from CSG-SP and CSG-FA or HAY-SP, but was greater in comparison to HAY-FA (p < 0.002).\n",
      "Span[1:2]: \"SCFA\" → Chemical (0.8532)\n",
      "********************\n",
      "Similarly, acetate was greater in CSG-FA than HAY-SP or HAY-FA, while WSG only differed from HAY-FA (p < 0.02).\n",
      "Span[2:3]: \"acetate\" → Chemical (0.904)\n",
      "********************\n",
      "Acetate concentrations for CSG-FA also differed with HAY-FA (p = 0.0003), but there was only a trend for a difference between CSG-SP and HAY-SP (p = 0.09).\n",
      "Span[0:1]: \"Acetate\" → Chemical (0.9293)\n",
      "********************\n",
      "Fecal butyrate was lower for HAY-FA than when horses were adapted to any of the pasture forages (p < 0.03), but there was no difference between HAY-SP and any other forage.\n",
      "Span[1:2]: \"butyrate\" → Chemical (0.9568)\n",
      "Span[8:9]: \"horses\" → Species (0.7168)\n",
      "********************\n",
      "Fecal propionate was lowest in horses adapted to HAY-FA (p < 0.002), but propionate concentrations were also lower for HAY-SP than either CSG-SP or CSG-FA (p < 0.03).\n",
      "Span[1:2]: \"propionate\" → Chemical (0.9449)\n",
      "Span[5:6]: \"horses\" → Species (0.9339)\n",
      "Span[16:17]: \"propionate\" → Chemical (0.9499)\n",
      "********************\n",
      "Propionate did not differ between horses adapted to WSG and all other forages.\n",
      "Span[0:1]: \"Propionate\" → Chemical (0.919)\n",
      "Span[5:6]: \"horses\" → Species (0.8991)\n",
      "********************\n",
      "Fecal valerate was lower for horses adapted to HAY-FA than CSG-SP or CSG-FA (p < 0.03) but did not differ from WSG or HAY-SP.\n",
      "Span[1:2]: \"valerate\" → Chemical (0.8694)\n",
      "Span[5:6]: \"horses\" → Species (0.8234)\n",
      "********************\n",
      "Horses adapted to HAY-SP also had fecal valerate concentrations lower than CSG-SP or CSG-FA (p ≤ 0.0002).\n",
      "Span[0:1]: \"Horses\" → Species (0.9317)\n",
      "Span[7:8]: \"valerate\" → Chemical (0.9052)\n",
      "********************\n",
      "Isobutyrate was greater for all pasture forages in comparison to both HAY-SP and HAY-FA (p < 0.02).\n",
      "Span[0:1]: \"Isobutyrate\" → Chemical (0.9504)\n",
      "********************\n",
      "Isovalerate was greater in CSG-SP and CSG-FA vs. HAY-SP and HAY-FA (p ≤ 0.002), while concentrations in horses adapted to WSG were once again intermediate.\n",
      "Span[0:1]: \"Isovalerate\" → Chemical (0.9601)\n",
      "Span[20:21]: \"horses\" → Species (0.9257)\n",
      "********************\n",
      "Isocaproate was detected in all fecal samples, but concentrations were negligible and below the limit of quantification (<1.0 ug g feces−1).\n",
      "Span[0:1]: \"Isocaproate\" → Chemical (0.97)\n",
      "********************\n",
      "Relatively weak predictive accuracy was found for isovalerate, valerate, hexanoate, and heptanoate (model R2 and p-values are shown in Table 7).\n",
      "Span[7:8]: \"isovalerate\" → Chemical (0.9498)\n",
      "Span[9:10]: \"valerate\" → Chemical (0.9548)\n",
      "Span[11:12]: \"hexanoate\" → Chemical (0.9617)\n",
      "Span[14:15]: \"heptanoate\" → Chemical (0.9283)\n",
      "********************\n",
      "Major SCFA including acetate, butyrate, and propionate could not be predicted based on BCG abundance profiles.\n",
      "Span[1:2]: \"SCFA\" → Chemical (0.9484)\n",
      "Span[3:4]: \"acetate\" → Chemical (0.961)\n",
      "Span[5:6]: \"butyrate\" → Chemical (0.9745)\n",
      "Span[8:9]: \"propionate\" → Chemical (0.9387)\n",
      "********************\n",
      "However, when random forest regressors were applied to only the top 25 BCG identified as most predictive of forage type, model accuracies improved for all fecal variables with the exception of valerate (Table 7).\n",
      "Span[33:34]: \"valerate\" → Chemical (0.9062)\n",
      "********************\n",
      "There were positive correlations between BCG_300 and total SCFA and total BCFA as well as with acetate, butyrate, propionate, valerate, isobutyrate, and isovalerate, while this BCG was negatively correlated with hexanoate, heptanoate, and pH.\n",
      "Span[10:11]: \"SCFA\" → Chemical (0.9694)\n",
      "Span[13:14]: \"BCFA\" → Chemical (0.987)\n",
      "Span[18:19]: \"acetate\" → Chemical (0.9385)\n",
      "Span[20:21]: \"butyrate\" → Chemical (0.9803)\n",
      "Span[22:23]: \"propionate\" → Chemical (0.9761)\n",
      "Span[24:25]: \"valerate\" → Chemical (0.9773)\n",
      "Span[26:27]: \"isobutyrate\" → Chemical (0.9627)\n",
      "Span[29:30]: \"isovalerate\" → Chemical (0.953)\n",
      "Span[38:39]: \"hexanoate\" → Chemical (0.9347)\n",
      "Span[40:41]: \"heptanoate\" → Chemical (0.9496)\n",
      "********************\n",
      "The ASV members of this BCG were assigned to the genera Lachnospiraceae XPB1014 and Streptococcus.\n",
      "Span[14:15]: \"Streptococcus\" → Species (0.7771)\n",
      "********************\n",
      "There were also positive correlations between both BCG_9 and BCG_259 and total SCFA and BCFA in addition to acetate, butyrate, propionate, valerate, isobutyrate, and isovalerate; these BCG were negatively correlated with heptoanate.\n",
      "Span[16:17]: \"SCFA\" → Chemical (0.9858)\n",
      "Span[18:19]: \"BCFA\" → Chemical (0.9927)\n",
      "Span[22:23]: \"acetate\" → Chemical (0.9331)\n",
      "Span[24:25]: \"butyrate\" → Chemical (0.9812)\n",
      "Span[26:27]: \"propionate\" → Chemical (0.9786)\n",
      "Span[28:29]: \"valerate\" → Chemical (0.9805)\n",
      "Span[30:31]: \"isobutyrate\" → Chemical (0.9661)\n",
      "Span[33:34]: \"isovalerate\" → Chemical (0.9603)\n",
      "Span[41:42]: \"heptoanate\" → Chemical (0.9253)\n",
      "********************\n",
      "These BCG included multiple ASV assigned to Lachnospiraceae as well as ASV within the genera Christensenellaceae R-7 group, Rikenellaceae RC9 gut group, Papillibacter, and the NK4214 group of Oscillospiraceae.\n",
      "Span[4:5]: \"ASV\" → Species (0.6397)\n",
      "Span[11:12]: \"ASV\" → Species (0.6306)\n",
      "********************\n",
      "Positive correlations were also found between BCG_173 and BCG_201 and total SCFA and BCFA, acetate, propionate, valerate, isobutyrate, and isovalerate, with negative correlations between these BCG and both hexoanate and heptoanate.\n",
      "Span[15:16]: \"SCFA\" → Chemical (0.9843)\n",
      "Span[17:18]: \"BCFA\" → Chemical (0.997)\n",
      "Span[19:20]: \"acetate\" → Chemical (0.9545)\n",
      "Span[21:22]: \"propionate\" → Chemical (0.9774)\n",
      "Span[23:24]: \"valerate\" → Chemical (0.979)\n",
      "Span[25:26]: \"isobutyrate\" → Chemical (0.9646)\n",
      "Span[28:29]: \"isovalerate\" → Chemical (0.9594)\n",
      "Span[35:36]: \"BCG\" → Chemical (0.7793)\n",
      "Span[38:39]: \"hexoanate\" → Chemical (0.9363)\n",
      "Span[40:41]: \"heptoanate\" → Chemical (0.9544)\n",
      "********************\n",
      "These BCG included ASV assigned to Treponema, Rikenellaceae RC9 gut group, the NKA214 group of Oscillospiraceae, and the p-251-o5 and F082 genera and families of Bacteroidales.\n",
      "Span[3:4]: \"ASV\" → Species (0.7001)\n",
      "********************\n",
      "Total SCFA and BCFA as well as acetate, butyrate, propionate, isobutyrate, and isovalerate were positively correlated with BCG_113, while hexanoate was negatively correlated.\n",
      "Span[1:2]: \"SCFA\" → Chemical (0.9723)\n",
      "Span[3:4]: \"BCFA\" → Chemical (0.9781)\n",
      "Span[7:8]: \"acetate\" → Chemical (0.9366)\n",
      "Span[9:10]: \"butyrate\" → Chemical (0.9758)\n",
      "Span[11:12]: \"propionate\" → Chemical (0.971)\n",
      "Span[13:14]: \"isobutyrate\" → Chemical (0.9627)\n",
      "Span[16:17]: \"isovalerate\" → Chemical (0.9407)\n",
      "Span[26:27]: \"hexanoate\" → Chemical (0.9399)\n",
      "********************\n",
      "This BCG contained ASV assigned to taxa including Synergistaceae, Bacteroidales RF16 group, Papillibacter, Christensenellaceae R-7 Group, Mogibacterium, and Clostridium butyricum.\n",
      "Span[3:4]: \"ASV\" → Species (0.5622)\n",
      "Span[23:25]: \"Clostridium butyricum\" → Species (0.7977)\n",
      "********************\n",
      "Seven BCG were positively correlated with total BCFA and some combination of propionate, valerate, isobutyrate, and isovalerate, while negative correlations were found between these BCG and hexanoate and/or heptanoate.\n",
      "Span[7:8]: \"BCFA\" → Chemical (0.9777)\n",
      "Span[12:13]: \"propionate\" → Chemical (0.9466)\n",
      "Span[14:15]: \"valerate\" → Chemical (0.9701)\n",
      "Span[16:17]: \"isobutyrate\" → Chemical (0.9546)\n",
      "Span[19:20]: \"isovalerate\" → Chemical (0.951)\n",
      "Span[28:29]: \"BCG\" → Chemical (0.8889)\n",
      "Span[30:31]: \"hexanoate\" → Chemical (0.8779)\n",
      "Span[34:35]: \"heptanoate\" → Chemical (0.9165)\n",
      "********************\n",
      "These BCG included ASV members assigned to the family level for Anaerovoracaceae, Oscillospiraceae, and Lachnospiraceae.\n",
      "Span[3:4]: \"ASV\" → Species (0.561)\n",
      "********************\n",
      "Other ASV within these BCG were assigned to Lactobacillus equigenerosi and the genera Akkermansia, Fibrobacter, Treponema, Sphaerocheata, Phasolarctobacterium, Christensenellaceae R-7 group, Lachnoclostridium, Cellulosilyticum, Marvinbryantia, Coprostanoligenes group, Lachnospiraceae XPB1014, Lachnospiraceae UCG-009, Erysipelatoclostridium, Clostridium sensu stricto 1, and Bacteroidales BS11 gut group as well as Family XIII AD3011 group of Anaerovoracaceae, the UCG-004 genus within Erysipelatoclostridiaceae, the UCG-002 and UCG-005 genuses of Oscillospiraceae, and the UCG-010 genus and family of Oscillospirales.\n",
      "Span[8:10]: \"Lactobacillus equigenerosi\" → Species (0.6602)\n",
      "********************\n",
      "This BCG included ASV assigned to Prevotella, Sarcina, and Clostridium sensu stricto 1.\n",
      "Span[3:4]: \"ASV\" → Species (0.6284)\n",
      "********************\n",
      "Application of random forest regressors demonstrated that forage nutrient concentrations could be predicted based on bacterial community composition (BCG composition of the full microbial community), including water-soluble carbohydrate (WSC) and NSC at a predictive accuracy > R2 = 0.50 and crude protein (CP) with a predictive accuracy of R2 = 0.41 (p < 0.0001; Table 7).\n",
      "Span[30:31]: \"carbohydrate\" → Chemical (0.9793)\n",
      "Span[35:36]: \"NSC\" → CellLine (0.8274)\n",
      "********************\n",
      "Similar to results for fecal metabolites, when random forest regressors were applied only to the top 25 BCG identified as most predictive of forage type, model accuracies improved (Table 7).\n",
      "Span[5:6]: \"metabolites\" → Chemical (0.7674)\n",
      "********************\n",
      "There were distinct ASV from the taxa Christensenellaceae R-7 group and NK4A214 group of Oscillospiraceae that were members of separate BCG which were positively and negatively responding to these nutrients (i.e., these taxonomic classifications were found across all positive and negative responder groups).\n",
      "Span[3:4]: \"ASV\" → Species (0.5328)\n",
      "********************\n",
      "The BCG negatively correlated with ADF/NDF, and thus positively correlated with CP/DE included ASV members assigned to Clostridium butyricum and genera Papillibacter, Lachnoclostridium, Cellulosilyticum, Fibrobacter, Treponema, Catenisphaera, Bacteroidales RF16 group, Synergistaceae, Mogibacterium, Rikenellaceae RC9 gut group, the Hallii group of Lachnospiraceae, the p-251-o5 genus and family of Oscillospiraceae, and the F082 genus and family of Bacteroidales, as well as multiple ASV assigned only to the family level for Lachnospiraceae.\n",
      "Span[18:19]: \"ASV\" → Species (0.4645)\n",
      "Span[22:24]: \"Clostridium butyricum\" → Species (0.847)\n",
      "Span[78:79]: \"ASV\" → Species (0.4972)\n",
      "********************\n",
      "The BCG positively correlated with ADF/NDF, and negatively correlated with CP/DE included ASV within Bacteroidales BS11 gut group and Clostridium sensu stricto 1.\n",
      "Span[24:27]: \"Clostridium sensu stricto\" → Species (0.6194)\n",
      "********************\n",
      "Forage ADF and NDF were also negatively correlated with BCG_300 (Streptococcus and Lachnospiraceae XPB1014), but this BCG was not correlated with either CP or DE.\n",
      "Span[13:14]: \"Streptococcus\" → Species (0.8052)\n",
      "********************\n",
      "The ASV members of BCG_35 were assigned to taxa including Akkermansia, Fibrobacter, Treponema, Christensenellaceae R-7 group, and Family XIII AD3011 within Anaerovoracaceae.\n",
      "Span[18:21]: \"Christensenellaceae R-7 group\" → Species (0.7608)\n",
      "Span[23:26]: \"Family XIII AD3011\" → Species (0.734)\n",
      "********************\n",
      "Forage NSC and WSC were correlated with 13 and 12 BCG, respectively (rs ≥ |0.30|; p ≤ 0.05).\n",
      "Span[1:2]: \"NSC\" → CellLine (0.7083)\n",
      "********************\n",
      "In addition to the above-mentioned taxa from ASV within BCG_35, ASV within BCG negatively correlated with both NSC and WSC were assigned to Lactobacillus equigenerosi and genera including Lachnoclostridium, Cellulosilyticum, Catenisphaera, Marvinbryantia, Erysipelatoclostridium, Mogibacterium, Lachnospiraceae XPB1014 group, Lachnospiraceae UCG-009, Coprostanoligenes group, UCG-005 metagenome within Oscillospiraceae, the Hallii group within Lachnospiraceae, the NK4A214 group of Oscillospiraceae, the Incertae Sedis genus of Ethanoligenenaceae, the WCHB1-41 genus, family, and order within Kiritimatiellae, as well as Denitrobacterium detoxificans.\n",
      "Span[91:93]: \"Denitrobacterium detoxificans\" → Species (0.6941)\n",
      "********************\n",
      "Forage NSC and WSC were positively correlated with BCG_94 (Clostridium sensu stricto 1, Prevotella, and Sarcina).\n",
      "Span[12:15]: \"Clostridium sensu stricto\" → Species (0.6416)\n",
      "********************\n",
      "Forage ESC was correlated with 9 of the same BCG as NSC and WSC, but also had a negative correlation with BCG_173, which contained multiple ASV assigned to Lachnospiraceae, probable genus 10 within Lachnospiraceae, Pseudobutyvibrio as well as to the p-251 genus and family of Bacteroidales.\n",
      "Span[22:25]: \"BCG_173\" → Species (0.6357)\n",
      "********************\n",
      "Acetate, butyrate, propionate, valerate, isobutyrate, isovalerate and total SCFA and BCFA were positively correlated with DE (Spearman correlation; rs ≥ 0.62) and CP (rs ≥ 0.41), but were negatively correlated with NDF (rs ≤ −0.48) and ADF (rs ≤ −0.57; p ≤ 0.008).\n",
      "Span[0:1]: \"Acetate\" → Chemical (0.9634)\n",
      "Span[2:3]: \"butyrate\" → Chemical (0.9781)\n",
      "Span[4:5]: \"propionate\" → Chemical (0.9698)\n",
      "Span[6:7]: \"valerate\" → Chemical (0.9681)\n",
      "Span[8:9]: \"isobutyrate\" → Chemical (0.9688)\n",
      "Span[10:11]: \"isovalerate\" → Chemical (0.9393)\n",
      "Span[13:14]: \"SCFA\" → Chemical (0.9571)\n",
      "Span[15:16]: \"BCFA\" → Chemical (0.9574)\n",
      "********************\n",
      "Total BCFA, isobutyrate, and isovalerate were negatively correlated with NSC, WSC, and ESC (rs ≤ −0.37; p ≤ 0.02), with a weaker negative correlation between valerate and NSC and WSC (rs ≤ −0.34; p ≤ 0.03).\n",
      "Span[1:2]: \"BCFA\" → Chemical (0.9924)\n",
      "Span[3:4]: \"isobutyrate\" → Chemical (0.9497)\n",
      "Span[6:7]: \"isovalerate\" → Chemical (0.9405)\n",
      "Span[34:35]: \"valerate\" → Chemical (0.9389)\n",
      "********************\n",
      "Hexanoate and heptanoate were negatively correlated with DE, CP, and starch (rs ≤ −0.39; p ≤ 0.01), but were positively correlated with NDF, ADF, NSC, WSC, and ESC (rs ≥ 0.37; p ≤ 0.02).\n",
      "Span[0:1]: \"Hexanoate\" → Chemical (0.9446)\n",
      "Span[2:3]: \"heptanoate\" → Chemical (0.916)\n",
      "********************\n",
      "Plasma glucose responses to OST administration differed by forage (mixed model ANOVA with Tukey’s post hoc adjustment; p ≤ 0.01).\n",
      "Span[1:2]: \"glucose\" → Chemical (0.9944)\n",
      "********************\n",
      "The AUC for glucose was lowest when horses were adapted to WSG (42.4 ± 6.8 mg/dL*h) vs. CSG-SP (70.0 ± 6.8 mg/dL*h) and HAY-FA (76.3 ± 6.8 mg/dL*h; p ≤ 0.03; Figure 7a).\n",
      "Span[3:4]: \"glucose\" → Chemical (0.9884)\n",
      "Span[7:8]: \"horses\" → Species (0.968)\n",
      "********************\n",
      "Peak plasma glucose was also lower for WSG (110 ± 3 mg/dL) in comparison to HAY-FA (123 ± 3 mg/dL; p = 0.01), and there was a trend for lower peak plasma glucose for WSG than for CSG-SP (120 ± 3 mg/dL; p = 0.06; Figure 7b).\n",
      "Span[2:3]: \"glucose\" → Chemical (0.9947)\n",
      "Span[42:43]: \"glucose\" → Chemical (0.9962)\n",
      "********************\n",
      "Forage did not, however, impact OST insulin responses.\n",
      "Span[8:9]: \"insulin\" → Gene (0.8972)\n",
      "********************\n",
      "Neither AUC (CSG-SP: 36.1; WSG: 29.1; HAY-FA: 32.7 ± 6.4 mIU/L*h) or peak plasma insulin (CSG-SP: 28.8; WSG: 25.5; HAY-FA: 26.6 ± 3.3 mIU/L) varied by forage (Figure S2).\n",
      "Span[25:26]: \"insulin\" → Gene (0.5557)\n",
      "********************\n",
      "Fasting plasma insulin (CSG-SP: 3.69; WSG: 5.04; HAY-FA: 5.20 ± 0.64 mIU/L) and glucose (CSG-SP: 81.4; WSG: 78.7; HAY-FA: 81.3 ± 1.4 mg/dL) also did not differ by forage (Figure S3).\n",
      "Span[2:3]: \"insulin\" → Gene (0.5653)\n",
      "Span[22:23]: \"glucose\" → Chemical (0.9896)\n",
      "********************\n",
      "There were trends for differences by forage in proxies for insulin sensitivity including the fasting glucose-to-insulin ratio (FGIR) and reciprocal of the square root of insulin (RISQI; mixed model ANOVA with Tukey’s post hoc adjustment; p ≤ 0.10; Figure S4).\n",
      "Span[10:12]: \"insulin sensitivity\" → Disease (0.6053)\n",
      "Span[10:11]: \"insulin\" → Gene (0.9635)\n",
      "Span[15:16]: \"glucose-to-insulin\" → Gene (0.9613)\n",
      "Span[27:28]: \"insulin\" → Gene (0.9114)\n",
      "********************\n",
      "However, the modified insulin-to-glucose ratio (MIRG), a proxy for the insulin secretory response, did not vary by forage (Figure S4).\n",
      "Span[4:5]: \"insulin-to-glucose\" → Chemical (0.9082)\n",
      "Span[14:15]: \"insulin\" → Gene (0.8769)\n",
      "********************\n",
      "Glucose and insulin dynamics were only assessed after three of the forages (CSG-SP, WSG, and HAY-FA), and this smaller dataset precluded the use of random forest regression modelling for these variables.\n",
      "Span[0:1]: \"Glucose\" → Chemical (0.9955)\n",
      "Span[2:3]: \"insulin\" → Gene (0.9326)\n",
      "********************\n",
      "As the primary metabolic differences in grazing horse metabolism were AUC and peak plasma glucose in response to OST administration, correlation analysis was conducted to explore relationships between these metabolic variables and BCG.\n",
      "Span[7:8]: \"horse\" → Species (0.7658)\n",
      "Span[14:15]: \"glucose\" → Chemical (0.9949)\n",
      "********************\n",
      "Only one BCG (of the 25 BCG in the reduced feature set) was correlated with AUC (rs = −0.48; p = 0.04), with members of BCG_124 including ASV within Papillibacter, Christensenellaceae R-7 group, and Synergistaceae.\n",
      "Span[32:35]: \"BCG_124\" → Species (0.6048)\n",
      "Span[36:37]: \"ASV\" → Species (0.7066)\n",
      "********************\n",
      "Four additional BCG were correlated with peak plasma glucose.\n",
      "Span[8:9]: \"glucose\" → Chemical (0.9915)\n",
      "********************\n",
      "Positive correlations were found between peak glucose and BCG_51 and BCG_305 (rs ≥ 0.43; p ≤ 0.04).\n",
      "Span[6:7]: \"glucose\" → Chemical (0.9896)\n",
      "********************\n",
      "These BCG included ASV mapped to genera including Lachnospiraceae XPB1014, Lachnospiraceae UCG-009, Erysipelatoclostridium, Catenisphaera, and Fibrobacter as well to the family level for Anaerovoracaceae.\n",
      "Span[3:4]: \"ASV\" → Species (0.5558)\n",
      "********************\n",
      "Conversely, BCG_113 (Clostridium butyricum, Papillibacter, Bacteroidales RF16 group) and BCG_259 were negatively correlated with peak glucose (rs = −0.44; p = 0.03).\n",
      "Span[6:8]: \"Clostridium butyricum\" → Species (0.7287)\n",
      "Span[24:25]: \"glucose\" → Chemical (0.9726)\n",
      "********************\n",
      "Members of BCG_259 included ASV assigned at the family level to Lachnospiraceae in addition to the Hallii group within Lachnospiraceae.\n",
      "Span[6:7]: \"ASV\" → Species (0.5467)\n",
      "********************\n",
      "While relationships between AUC and peak plasma glucose and forage nutrients were identified through correlation analysis (Spearman correlation; rs ≥ |0.41|; p ≤ 0.04; Figure 7c), these metabolic variables were not correlated with any fecal variables including SCFA, BCFA, and pH.\n",
      "Span[7:8]: \"glucose\" → Chemical (0.9939)\n",
      "Span[45:46]: \"SCFA\" → Chemical (0.9549)\n",
      "Span[47:48]: \"BCFA\" → Chemical (0.9878)\n",
      "********************\n",
      "Forage NSC, WSC, and ESC were positively correlated with AUC (rs ≥ 0.53; p ≤ 0.007) and peak glucose (rs ≥ 0.41; p ≤ 0.04).\n",
      "Span[1:2]: \"NSC\" → CellLine (0.6574)\n",
      "Span[23:24]: \"glucose\" → Chemical (0.9854)\n",
      "********************\n",
      "There was also a negative correlation between CP and glucose responses to OST administration (rs ≤ −0.50; p ≤ 0.01).\n",
      "Span[9:10]: \"glucose\" → Chemical (0.9886)\n",
      "********************\n",
      "Forage DE, NDF, ADF, and starch, however, were not correlated with AUC or peak glucose.\n",
      "Span[19:20]: \"glucose\" → Chemical (0.9774)\n",
      "********************\n",
      "Therefore, this study aimed of to characterize the fecal microbiota of horses adapted to different forage types and to explore relationships between forage nutrients, microbial composition, fecal metabolites, and glycemic responses of grazing horses.\n",
      "Span[12:13]: \"horses\" → Species (0.8242)\n",
      "Span[37:38]: \"horses\" → Species (0.5552)\n",
      "********************\n",
      "This indicates that distinct and identifiable shifts in microbial composition do occur as horses adapt to different forage types.\n",
      "Span[13:14]: \"horses\" → Species (0.5343)\n",
      "********************\n",
      "The BCG identified as markers specific to HAY-SP included multiple ASV assigned to taxa to which fibrolytic and butyrate-producing functions have been previously ascribed [74,75,76].\n",
      "Span[7:8]: \"HAY-SP\" → Disease (0.7398)\n",
      "********************\n",
      "The BCG specific to this forage contained a heavy representation of ASV within the Lachnospiraceae family as well as Fibrobacter, Ruminococcus, and Pseudobutyvibrio.\n",
      "Span[11:12]: \"ASV\" → Species (0.6055)\n",
      "********************\n",
      "Prior studies have also reported increased prevalence of Lachnospiraceae in horses fed hay vs. pasture [19,77].\n",
      "Span[10:11]: \"horses\" → Species (0.5271)\n",
      "********************\n",
      "Increases in this genus in response to abrupt inclusion of dietary starch have been documented [78] as well as in temporal proximity to oral administration of oligofructose in experimental laminitis induction models [79,80].\n",
      "Span[28:29]: \"oligofructose\" → Chemical (0.9553)\n",
      "Span[31:32]: \"laminitis\" → Disease (0.9792)\n",
      "********************\n",
      "The co-occurrence of bacteria in these taxonomic groups could potentially be reflective of the nutrient composition of HAY-SP, which was the highest in both fiber and NSC of all forages.\n",
      "Span[27:28]: \"NSC\" → CellLine (0.7267)\n",
      "********************\n",
      "These factors may also account for unexpected associations revealed by analysis in the current study, such as ASV assigned to Streptococcus within BCG markers of HAY-FA, despite the relatively low NSC content of this forage.\n",
      "Span[21:22]: \"Streptococcus\" → Species (0.7463)\n",
      "********************\n",
      "[77] also reported a lower abundance of species within Streptococcaceae in horses maintained on pasture vs. horses fed hay or silage [77].\n",
      "Span[13:14]: \"horses\" → Species (0.5135)\n",
      "Span[18:19]: \"horses\" → Species (0.7004)\n",
      "********************\n",
      "Overall, the fecal microbiota of horses adapted to cool-season pasture was characterized by bacteria capable of utilizing rapidly fermentable fibers, which could explain the greater fecal SCFA concentrations for CSG-SP and CSG-FA.\n",
      "Span[6:7]: \"horses\" → Species (0.7578)\n",
      "Span[28:29]: \"SCFA\" → Chemical (0.8207)\n",
      "********************\n",
      "Capacity for hemicellulose fermentation and SCFA production have been previously documented in the Bacteroidales BS11 gut group [81], which was identified only in BCG markers of CSG-SP.\n",
      "Span[5:6]: \"SCFA\" → Chemical (0.8623)\n",
      "********************\n",
      "Equine studies have found increased relative abundance of Bacteroidales BS11 gut group in horses fed barley vs. a hay diet [82] and in horses presenting with colic [83,84].\n",
      "Span[13:14]: \"horses\" → Species (0.6773)\n",
      "Span[15:16]: \"barley\" → Species (0.9451)\n",
      "Span[25:26]: \"horses\" → Species (0.7593)\n",
      "Span[28:29]: \"colic\" → Disease (0.9562)\n",
      "********************\n",
      "Prior studies have found enrichment of the phylum Kiritimatiellaeota in the fecal microbiota of horses with equine metabolic syndrome and insulin dysregulation [26,27].\n",
      "Span[14:15]: \"horses\" → Species (0.8095)\n",
      "Span[17:19]: \"metabolic syndrome\" → Disease (0.7369)\n",
      "Span[20:22]: \"insulin dysregulation\" → Disease (0.8765)\n",
      "Span[20:21]: \"insulin\" → Gene (0.8868)\n",
      "********************\n",
      "[27] reported increased abundance of this phylum in response to a change from pasture to a hay diet in both healthy and insulin dysregulated ponies, and Ericsson et al.\n",
      "Span[24:25]: \"insulin\" → Gene (0.9268)\n",
      "********************\n",
      "[88] similarly found increased abundance of the class Kiritimatiellae when horses with equine metabolic syndrome were transitioned from pasture to a hay diet.\n",
      "Span[12:13]: \"horses\" → Species (0.7304)\n",
      "Span[15:17]: \"metabolic syndrome\" → Disease (0.701)\n",
      "********************\n",
      "A number of interesting relationships were found between BCG identified as markers specific to WSG and fecal variables, forage nutrients, and horse glucose metabolism.\n",
      "Span[23:24]: \"horse\" → Species (0.945)\n",
      "Span[24:25]: \"glucose\" → Chemical (0.9781)\n",
      "********************\n",
      "Warm-season grasses are characteristically lower in soluble carbohydrates than cool-season grasses [1,5], and of all forages evaluated in the current study, NSC and WSC were lowest in WSG.\n",
      "Span[7:8]: \"carbohydrates\" → Chemical (0.9859)\n",
      "********************\n",
      "Four of the five BCG markers of WSG were negatively correlated with forage NSC and WSC, including BCG_35, which contained ASV assigned Akkermansia as well as Fibrobacter, Treponema, Christensenellaceae R-7 group, and the Family XIII AD3011 group of Anaerovoracaceae.\n",
      "Span[18:21]: \"BCG_35\" → Species (0.6109)\n",
      "Span[24:25]: \"ASV\" → Species (0.5897)\n",
      "********************\n",
      "[93] recently demonstrated immune (and inflammatory) modulation by specific hindgut microbiota in horses, including Akkermansia spp.\n",
      "Span[16:17]: \"horses\" → Species (0.8374)\n",
      "********************\n",
      "Furthermore, supplementation with mannanoligosaccharides and fructooligosaccharides led to increased abundance of Akkermansia muciniphilia in a subsequent study [94], and improvements in insulin sensitivity have been previously documented in horses supplemented with low doses of fructooligosaccharide [95,96].\n",
      "Span[4:5]: \"mannanoligosaccharides\" → Chemical (0.9562)\n",
      "Span[6:7]: \"fructooligosaccharides\" → Chemical (0.9417)\n",
      "Span[12:14]: \"Akkermansia muciniphilia\" → Species (0.7246)\n",
      "Span[25:27]: \"insulin sensitivity\" → Disease (0.787)\n",
      "Span[25:26]: \"insulin\" → Gene (0.9565)\n",
      "Span[32:33]: \"horses\" → Species (0.9554)\n",
      "Span[38:39]: \"fructooligosaccharide\" → Chemical (0.8803)\n",
      "********************\n",
      "Markers of WSG also included BCG_113, which contained an ASV assigned to Clostridium butyricum.\n",
      "Span[15:17]: \"Clostridium butyricum\" → Species (0.7302)\n",
      "********************\n",
      "Clostridium butyricum is a prolific butyrate producer that has also been investigated for probiotic use due to its capacity to promote anti-inflammatory responses and improve gut barrier function and metabolic health [97,98,99].\n",
      "Span[0:2]: \"Clostridium butyricum\" → Species (0.8303)\n",
      "Span[5:6]: \"butyrate\" → Chemical (0.919)\n",
      "********************\n",
      "BCG_35 was positively correlated with CP as well as isobutyrate and total BCFA, which reflects the proteolytic capacity of this BCG.\n",
      "Span[11:12]: \"isobutyrate\" → Chemical (0.9416)\n",
      "Span[14:15]: \"BCFA\" → Chemical (0.9905)\n",
      "********************\n",
      "BCG_113 was also positively correlated with total SCFA and all three major SCFA (acetate, butyrate, and propionate) in addition to isobutyrate, isovalerate, and total BCFA.\n",
      "Span[9:10]: \"SCFA\" → Chemical (0.9377)\n",
      "Span[14:15]: \"SCFA\" → Chemical (0.9029)\n",
      "Span[16:17]: \"acetate\" → Chemical (0.9364)\n",
      "Span[18:19]: \"butyrate\" → Chemical (0.9743)\n",
      "Span[21:22]: \"propionate\" → Chemical (0.9556)\n",
      "Span[26:27]: \"isobutyrate\" → Chemical (0.9611)\n",
      "Span[28:29]: \"isovalerate\" → Chemical (0.9498)\n",
      "Span[32:33]: \"BCFA\" → Chemical (0.9836)\n",
      "********************\n",
      "This BCG was also negatively correlated with peak plasma glucose.\n",
      "Span[9:10]: \"glucose\" → Chemical (0.989)\n",
      "********************\n",
      "These findings suggest that this bacterial group including Clostridium butyricum could play a role in modulation of equine metabolic health.\n",
      "Span[8:10]: \"Clostridium butyricum\" → Species (0.7319)\n",
      "********************\n",
      "While limited connections between the gut microbiota and metabolic responses were observed in the current study, three of the five BCG correlated with AUC and peak plasma glucose were WSG-specific BCG markers.\n",
      "Span[28:29]: \"glucose\" → Chemical (0.9933)\n",
      "********************\n",
      "While Akkermansia and Clostridium butyricum have been investigated in mice and other species, comparatively little research has been conducted to understand the function of these bacteria in the equine hindgut.\n",
      "Span[3:5]: \"Clostridium butyricum\" → Species (0.7191)\n",
      "Span[9:10]: \"mice\" → Species (0.8609)\n",
      "********************\n",
      "Results of this study also confirmed the strong influence of dietary nutrients on the equine microbiome and provided insights into the complex relationships between forage nutrients and the gut microbiota.\n",
      "Span[14:15]: \"equine\" → Species (0.5663)\n",
      "********************\n",
      "This indicates that differences in forage nutrients were driving the shifts in equine fecal microbial communities as horses adapted to various forages within the integrated grazing system.\n",
      "Span[17:18]: \"horses\" → Species (0.6203)\n",
      "********************\n",
      "Results of this study revealed the influence of soluble carbohydrates including NSC and WSC on fecal microbial community composition.\n",
      "Span[9:10]: \"carbohydrates\" → Chemical (0.9803)\n",
      "********************\n",
      "Random forest modeling also revealed that in addition to soluble carbohydrates, the gut microbiota were also influenced by CP (R2 = 0.62).\n",
      "Span[10:11]: \"carbohydrates\" → Chemical (0.9706)\n",
      "********************\n",
      "The impact of CP on the gut microbial community was interesting, but unsurprising, as two-thirds to three-fourths of total tract nitrogen digestion and absorption occurs in the equine hindgut [100], and many microbial species are capable of proteolytic fermentation [101].\n",
      "Span[22:23]: \"nitrogen\" → Chemical (0.9572)\n",
      "Span[29:30]: \"equine\" → Species (0.6252)\n",
      "********************\n",
      "Prior studies in horses have found distinct effects of high-fiber vs. lower-fiber diets on the hindgut microbial community, with benefits of increased fiber including greater microbial diversity [102], reduction of lactic-acid producing bacteria [101,103], and less-acidic hindgut pH [104,105].\n",
      "Span[3:4]: \"horses\" → Species (0.9022)\n",
      "Span[34:35]: \"lactic-acid\" → Chemical (0.7131)\n",
      "********************\n",
      "However, these previous studies have been primarily conducted in horses fed concentrate vs. forage-based diets in which there was broad variation in fiber concentrations between treatments.\n",
      "Span[10:11]: \"horses\" → Species (0.9455)\n",
      "********************\n",
      "In the current study, horses were maintained on a forage-only diet throughout, and there were relatively high concentrations of NDF and ADF found across all forages.\n",
      "Span[5:6]: \"horses\" → Species (0.9693)\n",
      "********************\n",
      "Additionally, in contrast to fiber, these nutrients are all subject to digestion in the equine foregut, and only a fraction of the total soluble carbohydrates and CP ingested would be available for bacterial fermentation in the hindgut.\n",
      "Span[16:17]: \"equine\" → Species (0.7852)\n",
      "Span[27:28]: \"carbohydrates\" → Chemical (0.9673)\n",
      "Span[35:36]: \"bacterial\" → Species (0.8964)\n",
      "********************\n",
      "Fecal pH is commonly utilized as a marker of microbial activity in the equine hindgut [20,103].\n",
      "Span[13:14]: \"equine\" → Species (0.6349)\n",
      "********************\n",
      "Differences in fecal pH in the current study indicated that functional changes occur in the equine gut microbiota, in addition to shifts in microbiome structure and composition, as horses adapt to different forages.\n",
      "Span[30:31]: \"horses\" → Species (0.6002)\n",
      "********************\n",
      "Fecal pH was highest in horses adapted to WSG and HAY-FA, which were lower in NSC than CSG-SP, CSG-FA.\n",
      "Span[5:6]: \"horses\" → Species (0.8874)\n",
      "Span[16:17]: \"NSC\" → CellLine (0.845)\n",
      "********************\n",
      "Lower fecal pH is broadly associated with hindgut dysfunction in horses [106,107], and thus the higher fecal pH in horses adapted to WSG in comparison to cool-season pasture could suggest some benefit to grazing horses on warm-season grass pastures.\n",
      "Span[7:9]: \"hindgut dysfunction\" → Disease (0.7499)\n",
      "Span[22:23]: \"horses\" → Species (0.6178)\n",
      "********************\n",
      "However, mean pH was above 6.5 for CSG-SP, CSG-FA, and HAY-SP, which is within ranges previously documented in healthy forage-fed horses [33,108].\n",
      "Span[24:25]: \"horses\" → Species (0.6971)\n",
      "********************\n",
      "Therefore, while statistically significant, the difference in fecal pH between horses adapted to WSG vs. cool-season pasture may not be of physiological relevance.\n",
      "Span[12:13]: \"horses\" → Species (0.6619)\n",
      "********************\n",
      "Differences in fermentation metabolites across forages did not mirror results for fecal pH, as differences between forages for SCFA and BCFA were primarily between pasture forages and the hay diets.\n",
      "Span[19:20]: \"SCFA\" → Chemical (0.9595)\n",
      "Span[21:22]: \"BCFA\" → Chemical (0.9726)\n",
      "********************\n",
      "Numerically, SCFA and BCFA concentrations in horses adapted to WSG were intermediate between the cool-season pasture (greatest) and hay (lowest), but these differences were not statistically significant.\n",
      "Span[2:3]: \"SCFA\" → Chemical (0.9554)\n",
      "Span[4:5]: \"BCFA\" → Chemical (0.9875)\n",
      "Span[7:8]: \"horses\" → Species (0.9584)\n",
      "********************\n",
      "While there were distinct shifts in the fecal microbiota across forages in the integrated system, forage type exerted minimal effects on glucose metabolism.\n",
      "Span[22:23]: \"glucose\" → Chemical (0.9696)\n",
      "********************\n",
      "A large body of research, primarily conducted in mouse models, has established that diet (and dietary intervention) modulates metabolic health in a microbiome-dependent manner [29,30,114].\n",
      "Span[9:10]: \"mouse\" → Species (0.9275)\n",
      "********************\n",
      "The adaptation period utilized in the present study was of a similar duration as used in prior studies evaluating the microbiome of forage-fed horses [22,23,116], and is considered sufficient for stabilization of microbial communities [95].\n",
      "Span[23:24]: \"horses\" → Species (0.5491)\n",
      "********************\n",
      "The lower AUC and peak plasma glucose in horses adapted to WSG in the current study does suggest that glucose may be cleared more rapidly in horses adapted to this forage, but a longer adaptation period would be necessary to confirm these results.\n",
      "Span[6:7]: \"glucose\" → Chemical (0.9932)\n",
      "Span[8:9]: \"horses\" → Species (0.9793)\n",
      "Span[19:20]: \"glucose\" → Chemical (0.9896)\n",
      "Span[26:27]: \"horses\" → Species (0.9126)\n",
      "********************\n",
      "Regardless, glucose and insulin responses to the OST for all forages evaluated in the current study were within normal and previously reported ranges [12,34], indicating that substantial metabolic changes are unlikely within the context of integrated rotational grazing management even with the observed changes in the fecal microbiota, fecal metabolites, and pH.\n",
      "Span[2:3]: \"glucose\" → Chemical (0.9915)\n",
      "Span[4:5]: \"insulin\" → Gene (0.9314)\n",
      "********************\n",
      "Additionally, the lack of correlation between any fermentation metabolites and glucose responses, as well as the relatively small number of BCG correlated with AUC and peak plasma glucose, suggest that any difference in glycemic responses of horses in the current study may not have been heavily dependent on shifts in the hindgut microbiome.\n",
      "Span[11:12]: \"glucose\" → Chemical (0.9836)\n",
      "Span[29:30]: \"glucose\" → Chemical (0.9922)\n",
      "Span[39:40]: \"horses\" → Species (0.8622)\n",
      "********************\n",
      "Seasonal variance in insulin sensitivity has been documented in grazing horses [15,122], but when fed controlled diets without nutrient fluctuations seen in pasture forages, minimal seasonal differences in circulating glucose and insulin and insulin sensitivity have been found in healthy horses [15,123,124].\n",
      "Span[3:4]: \"insulin\" → Gene (0.965)\n",
      "Span[10:11]: \"horses\" → Species (0.7909)\n",
      "Span[33:34]: \"glucose\" → Chemical (0.9967)\n",
      "Span[35:36]: \"insulin\" → Gene (0.7956)\n",
      "Span[37:38]: \"insulin\" → Gene (0.9256)\n",
      "Span[44:45]: \"horses\" → Species (0.673)\n",
      "********************\n",
      "Therefore, this study aimed of to characterize the fecal microbiota of horses adapted to different forage types and to explore relationships between forage nutrients, microbial composition, fecal metabolites, and glycemic responses of grazing horses.\n",
      "Span[12:13]: \"horses\" → Species (0.8242)\n",
      "Span[37:38]: \"horses\" → Species (0.5552)\n",
      "********************\n",
      "This indicates that distinct and identifiable shifts in microbial composition do occur as horses adapt to different forage types.\n",
      "Span[13:14]: \"horses\" → Species (0.5343)\n",
      "********************\n",
      "The BCG identified as markers specific to HAY-SP included multiple ASV assigned to taxa to which fibrolytic and butyrate-producing functions have been previously ascribed [74,75,76].\n",
      "Span[7:8]: \"HAY-SP\" → Disease (0.7398)\n",
      "********************\n",
      "The BCG specific to this forage contained a heavy representation of ASV within the Lachnospiraceae family as well as Fibrobacter, Ruminococcus, and Pseudobutyvibrio.\n",
      "Span[11:12]: \"ASV\" → Species (0.6055)\n",
      "********************\n",
      "Prior studies have also reported increased prevalence of Lachnospiraceae in horses fed hay vs. pasture [19,77].\n",
      "Span[10:11]: \"horses\" → Species (0.5271)\n",
      "********************\n",
      "Increases in this genus in response to abrupt inclusion of dietary starch have been documented [78] as well as in temporal proximity to oral administration of oligofructose in experimental laminitis induction models [79,80].\n",
      "Span[28:29]: \"oligofructose\" → Chemical (0.9553)\n",
      "Span[31:32]: \"laminitis\" → Disease (0.9792)\n",
      "********************\n",
      "The co-occurrence of bacteria in these taxonomic groups could potentially be reflective of the nutrient composition of HAY-SP, which was the highest in both fiber and NSC of all forages.\n",
      "Span[27:28]: \"NSC\" → CellLine (0.7267)\n",
      "********************\n",
      "These factors may also account for unexpected associations revealed by analysis in the current study, such as ASV assigned to Streptococcus within BCG markers of HAY-FA, despite the relatively low NSC content of this forage.\n",
      "Span[21:22]: \"Streptococcus\" → Species (0.7463)\n",
      "********************\n",
      "[77] also reported a lower abundance of species within Streptococcaceae in horses maintained on pasture vs. horses fed hay or silage [77].\n",
      "Span[13:14]: \"horses\" → Species (0.5135)\n",
      "Span[18:19]: \"horses\" → Species (0.7004)\n",
      "********************\n",
      "Overall, the fecal microbiota of horses adapted to cool-season pasture was characterized by bacteria capable of utilizing rapidly fermentable fibers, which could explain the greater fecal SCFA concentrations for CSG-SP and CSG-FA.\n",
      "Span[6:7]: \"horses\" → Species (0.7578)\n",
      "Span[28:29]: \"SCFA\" → Chemical (0.8207)\n",
      "********************\n",
      "Capacity for hemicellulose fermentation and SCFA production have been previously documented in the Bacteroidales BS11 gut group [81], which was identified only in BCG markers of CSG-SP.\n",
      "Span[5:6]: \"SCFA\" → Chemical (0.8623)\n",
      "********************\n",
      "Equine studies have found increased relative abundance of Bacteroidales BS11 gut group in horses fed barley vs. a hay diet [82] and in horses presenting with colic [83,84].\n",
      "Span[13:14]: \"horses\" → Species (0.6773)\n",
      "Span[15:16]: \"barley\" → Species (0.9451)\n",
      "Span[25:26]: \"horses\" → Species (0.7593)\n",
      "Span[28:29]: \"colic\" → Disease (0.9562)\n",
      "********************\n",
      "Prior studies have found enrichment of the phylum Kiritimatiellaeota in the fecal microbiota of horses with equine metabolic syndrome and insulin dysregulation [26,27].\n",
      "Span[14:15]: \"horses\" → Species (0.8095)\n",
      "Span[17:19]: \"metabolic syndrome\" → Disease (0.7369)\n",
      "Span[20:22]: \"insulin dysregulation\" → Disease (0.8765)\n",
      "Span[20:21]: \"insulin\" → Gene (0.8868)\n",
      "********************\n",
      "[27] reported increased abundance of this phylum in response to a change from pasture to a hay diet in both healthy and insulin dysregulated ponies, and Ericsson et al.\n",
      "Span[24:25]: \"insulin\" → Gene (0.9268)\n",
      "********************\n",
      "[88] similarly found increased abundance of the class Kiritimatiellae when horses with equine metabolic syndrome were transitioned from pasture to a hay diet.\n",
      "Span[12:13]: \"horses\" → Species (0.7304)\n",
      "Span[15:17]: \"metabolic syndrome\" → Disease (0.701)\n",
      "********************\n",
      "A number of interesting relationships were found between BCG identified as markers specific to WSG and fecal variables, forage nutrients, and horse glucose metabolism.\n",
      "Span[23:24]: \"horse\" → Species (0.945)\n",
      "Span[24:25]: \"glucose\" → Chemical (0.9781)\n",
      "********************\n",
      "Warm-season grasses are characteristically lower in soluble carbohydrates than cool-season grasses [1,5], and of all forages evaluated in the current study, NSC and WSC were lowest in WSG.\n",
      "Span[7:8]: \"carbohydrates\" → Chemical (0.9859)\n",
      "********************\n",
      "Four of the five BCG markers of WSG were negatively correlated with forage NSC and WSC, including BCG_35, which contained ASV assigned Akkermansia as well as Fibrobacter, Treponema, Christensenellaceae R-7 group, and the Family XIII AD3011 group of Anaerovoracaceae.\n",
      "Span[18:21]: \"BCG_35\" → Species (0.6109)\n",
      "Span[24:25]: \"ASV\" → Species (0.5897)\n",
      "********************\n",
      "[93] recently demonstrated immune (and inflammatory) modulation by specific hindgut microbiota in horses, including Akkermansia spp.\n",
      "Span[16:17]: \"horses\" → Species (0.8374)\n",
      "********************\n",
      "Furthermore, supplementation with mannanoligosaccharides and fructooligosaccharides led to increased abundance of Akkermansia muciniphilia in a subsequent study [94], and improvements in insulin sensitivity have been previously documented in horses supplemented with low doses of fructooligosaccharide [95,96].\n",
      "Span[4:5]: \"mannanoligosaccharides\" → Chemical (0.9562)\n",
      "Span[6:7]: \"fructooligosaccharides\" → Chemical (0.9417)\n",
      "Span[12:14]: \"Akkermansia muciniphilia\" → Species (0.7246)\n",
      "Span[25:27]: \"insulin sensitivity\" → Disease (0.787)\n",
      "Span[25:26]: \"insulin\" → Gene (0.9565)\n",
      "Span[32:33]: \"horses\" → Species (0.9554)\n",
      "Span[38:39]: \"fructooligosaccharide\" → Chemical (0.8803)\n",
      "********************\n",
      "Markers of WSG also included BCG_113, which contained an ASV assigned to Clostridium butyricum.\n",
      "Span[15:17]: \"Clostridium butyricum\" → Species (0.7302)\n",
      "********************\n",
      "Clostridium butyricum is a prolific butyrate producer that has also been investigated for probiotic use due to its capacity to promote anti-inflammatory responses and improve gut barrier function and metabolic health [97,98,99].\n",
      "Span[0:2]: \"Clostridium butyricum\" → Species (0.8303)\n",
      "Span[5:6]: \"butyrate\" → Chemical (0.919)\n",
      "********************\n",
      "BCG_35 was positively correlated with CP as well as isobutyrate and total BCFA, which reflects the proteolytic capacity of this BCG.\n",
      "Span[11:12]: \"isobutyrate\" → Chemical (0.9416)\n",
      "Span[14:15]: \"BCFA\" → Chemical (0.9905)\n",
      "********************\n",
      "BCG_113 was also positively correlated with total SCFA and all three major SCFA (acetate, butyrate, and propionate) in addition to isobutyrate, isovalerate, and total BCFA.\n",
      "Span[9:10]: \"SCFA\" → Chemical (0.9377)\n",
      "Span[14:15]: \"SCFA\" → Chemical (0.9029)\n",
      "Span[16:17]: \"acetate\" → Chemical (0.9364)\n",
      "Span[18:19]: \"butyrate\" → Chemical (0.9743)\n",
      "Span[21:22]: \"propionate\" → Chemical (0.9556)\n",
      "Span[26:27]: \"isobutyrate\" → Chemical (0.9611)\n",
      "Span[28:29]: \"isovalerate\" → Chemical (0.9498)\n",
      "Span[32:33]: \"BCFA\" → Chemical (0.9836)\n",
      "********************\n",
      "This BCG was also negatively correlated with peak plasma glucose.\n",
      "Span[9:10]: \"glucose\" → Chemical (0.989)\n",
      "********************\n",
      "These findings suggest that this bacterial group including Clostridium butyricum could play a role in modulation of equine metabolic health.\n",
      "Span[8:10]: \"Clostridium butyricum\" → Species (0.7319)\n",
      "********************\n",
      "While limited connections between the gut microbiota and metabolic responses were observed in the current study, three of the five BCG correlated with AUC and peak plasma glucose were WSG-specific BCG markers.\n",
      "Span[28:29]: \"glucose\" → Chemical (0.9933)\n",
      "********************\n",
      "While Akkermansia and Clostridium butyricum have been investigated in mice and other species, comparatively little research has been conducted to understand the function of these bacteria in the equine hindgut.\n",
      "Span[3:5]: \"Clostridium butyricum\" → Species (0.7191)\n",
      "Span[9:10]: \"mice\" → Species (0.8609)\n",
      "********************\n",
      "Results of this study also confirmed the strong influence of dietary nutrients on the equine microbiome and provided insights into the complex relationships between forage nutrients and the gut microbiota.\n",
      "Span[14:15]: \"equine\" → Species (0.5663)\n",
      "********************\n",
      "This indicates that differences in forage nutrients were driving the shifts in equine fecal microbial communities as horses adapted to various forages within the integrated grazing system.\n",
      "Span[17:18]: \"horses\" → Species (0.6203)\n",
      "********************\n",
      "Results of this study revealed the influence of soluble carbohydrates including NSC and WSC on fecal microbial community composition.\n",
      "Span[9:10]: \"carbohydrates\" → Chemical (0.9803)\n",
      "********************\n",
      "Random forest modeling also revealed that in addition to soluble carbohydrates, the gut microbiota were also influenced by CP (R2 = 0.62).\n",
      "Span[10:11]: \"carbohydrates\" → Chemical (0.9706)\n",
      "********************\n",
      "The impact of CP on the gut microbial community was interesting, but unsurprising, as two-thirds to three-fourths of total tract nitrogen digestion and absorption occurs in the equine hindgut [100], and many microbial species are capable of proteolytic fermentation [101].\n",
      "Span[22:23]: \"nitrogen\" → Chemical (0.9572)\n",
      "Span[29:30]: \"equine\" → Species (0.6252)\n",
      "********************\n",
      "Prior studies in horses have found distinct effects of high-fiber vs. lower-fiber diets on the hindgut microbial community, with benefits of increased fiber including greater microbial diversity [102], reduction of lactic-acid producing bacteria [101,103], and less-acidic hindgut pH [104,105].\n",
      "Span[3:4]: \"horses\" → Species (0.9022)\n",
      "Span[34:35]: \"lactic-acid\" → Chemical (0.7131)\n",
      "********************\n",
      "However, these previous studies have been primarily conducted in horses fed concentrate vs. forage-based diets in which there was broad variation in fiber concentrations between treatments.\n",
      "Span[10:11]: \"horses\" → Species (0.9455)\n",
      "********************\n",
      "In the current study, horses were maintained on a forage-only diet throughout, and there were relatively high concentrations of NDF and ADF found across all forages.\n",
      "Span[5:6]: \"horses\" → Species (0.9693)\n",
      "********************\n",
      "Additionally, in contrast to fiber, these nutrients are all subject to digestion in the equine foregut, and only a fraction of the total soluble carbohydrates and CP ingested would be available for bacterial fermentation in the hindgut.\n",
      "Span[16:17]: \"equine\" → Species (0.7852)\n",
      "Span[27:28]: \"carbohydrates\" → Chemical (0.9673)\n",
      "Span[35:36]: \"bacterial\" → Species (0.8964)\n",
      "********************\n",
      "Fecal pH is commonly utilized as a marker of microbial activity in the equine hindgut [20,103].\n",
      "Span[13:14]: \"equine\" → Species (0.6349)\n",
      "********************\n",
      "Differences in fecal pH in the current study indicated that functional changes occur in the equine gut microbiota, in addition to shifts in microbiome structure and composition, as horses adapt to different forages.\n",
      "Span[30:31]: \"horses\" → Species (0.6002)\n",
      "********************\n",
      "Fecal pH was highest in horses adapted to WSG and HAY-FA, which were lower in NSC than CSG-SP, CSG-FA.\n",
      "Span[5:6]: \"horses\" → Species (0.8874)\n",
      "Span[16:17]: \"NSC\" → CellLine (0.845)\n",
      "********************\n",
      "Lower fecal pH is broadly associated with hindgut dysfunction in horses [106,107], and thus the higher fecal pH in horses adapted to WSG in comparison to cool-season pasture could suggest some benefit to grazing horses on warm-season grass pastures.\n",
      "Span[7:9]: \"hindgut dysfunction\" → Disease (0.7499)\n",
      "Span[22:23]: \"horses\" → Species (0.6178)\n",
      "********************\n",
      "However, mean pH was above 6.5 for CSG-SP, CSG-FA, and HAY-SP, which is within ranges previously documented in healthy forage-fed horses [33,108].\n",
      "Span[24:25]: \"horses\" → Species (0.6971)\n",
      "********************\n",
      "Therefore, while statistically significant, the difference in fecal pH between horses adapted to WSG vs. cool-season pasture may not be of physiological relevance.\n",
      "Span[12:13]: \"horses\" → Species (0.6619)\n",
      "********************\n",
      "Differences in fermentation metabolites across forages did not mirror results for fecal pH, as differences between forages for SCFA and BCFA were primarily between pasture forages and the hay diets.\n",
      "Span[19:20]: \"SCFA\" → Chemical (0.9595)\n",
      "Span[21:22]: \"BCFA\" → Chemical (0.9726)\n",
      "********************\n",
      "Numerically, SCFA and BCFA concentrations in horses adapted to WSG were intermediate between the cool-season pasture (greatest) and hay (lowest), but these differences were not statistically significant.\n",
      "Span[2:3]: \"SCFA\" → Chemical (0.9554)\n",
      "Span[4:5]: \"BCFA\" → Chemical (0.9875)\n",
      "Span[7:8]: \"horses\" → Species (0.9584)\n",
      "********************\n",
      "While there were distinct shifts in the fecal microbiota across forages in the integrated system, forage type exerted minimal effects on glucose metabolism.\n",
      "Span[22:23]: \"glucose\" → Chemical (0.9696)\n",
      "********************\n",
      "A large body of research, primarily conducted in mouse models, has established that diet (and dietary intervention) modulates metabolic health in a microbiome-dependent manner [29,30,114].\n",
      "Span[9:10]: \"mouse\" → Species (0.9275)\n",
      "********************\n",
      "The adaptation period utilized in the present study was of a similar duration as used in prior studies evaluating the microbiome of forage-fed horses [22,23,116], and is considered sufficient for stabilization of microbial communities [95].\n",
      "Span[23:24]: \"horses\" → Species (0.5491)\n",
      "********************\n",
      "The lower AUC and peak plasma glucose in horses adapted to WSG in the current study does suggest that glucose may be cleared more rapidly in horses adapted to this forage, but a longer adaptation period would be necessary to confirm these results.\n",
      "Span[6:7]: \"glucose\" → Chemical (0.9932)\n",
      "Span[8:9]: \"horses\" → Species (0.9793)\n",
      "Span[19:20]: \"glucose\" → Chemical (0.9896)\n",
      "Span[26:27]: \"horses\" → Species (0.9126)\n",
      "********************\n",
      "Regardless, glucose and insulin responses to the OST for all forages evaluated in the current study were within normal and previously reported ranges [12,34], indicating that substantial metabolic changes are unlikely within the context of integrated rotational grazing management even with the observed changes in the fecal microbiota, fecal metabolites, and pH.\n",
      "Span[2:3]: \"glucose\" → Chemical (0.9915)\n",
      "Span[4:5]: \"insulin\" → Gene (0.9314)\n",
      "********************\n",
      "Additionally, the lack of correlation between any fermentation metabolites and glucose responses, as well as the relatively small number of BCG correlated with AUC and peak plasma glucose, suggest that any difference in glycemic responses of horses in the current study may not have been heavily dependent on shifts in the hindgut microbiome.\n",
      "Span[11:12]: \"glucose\" → Chemical (0.9836)\n",
      "Span[29:30]: \"glucose\" → Chemical (0.9922)\n",
      "Span[39:40]: \"horses\" → Species (0.8622)\n",
      "********************\n",
      "Seasonal variance in insulin sensitivity has been documented in grazing horses [15,122], but when fed controlled diets without nutrient fluctuations seen in pasture forages, minimal seasonal differences in circulating glucose and insulin and insulin sensitivity have been found in healthy horses [15,123,124].\n",
      "Span[3:4]: \"insulin\" → Gene (0.965)\n",
      "Span[10:11]: \"horses\" → Species (0.7909)\n",
      "Span[33:34]: \"glucose\" → Chemical (0.9967)\n",
      "Span[35:36]: \"insulin\" → Gene (0.7956)\n",
      "Span[37:38]: \"insulin\" → Gene (0.9256)\n",
      "Span[44:45]: \"horses\" → Species (0.673)\n",
      "********************\n",
      "In conclusion, distinct shifts in equine fecal microbial community structure and composition occur as horses adapt to different forages within an integrated warm- and cool-season grass rotational pasture system, but a substantial impact of this management practice on glucose metabolism in healthy adult grazing horses is unlikely.\n",
      "Span[41:42]: \"glucose\" → Chemical (0.9796)\n",
      "********************\n",
      "Results of this study underscore the potential for relatively small amounts of NSC to influence hindgut microbial composition and also that protein utilization may be an important ecological niche within the microbiome of forage-fed horses.\n",
      "Span[12:13]: \"NSC\" → CellLine (0.8271)\n",
      "********************\n",
      "Fecal BCFA and SCFA concentrations were higher in horses adapted to all pasture forages versus hay, but in comparison to forage nutrients, bacterial community composition did not have as strong an impact on fermentation metabolites, likely reflecting functional redundancy of the microbial community.\n",
      "Span[1:2]: \"BCFA\" → Chemical (0.9838)\n",
      "Span[3:4]: \"SCFA\" → Chemical (0.9304)\n",
      "Span[8:9]: \"horses\" → Species (0.7696)\n",
      "********************\n",
      "The guild-based analytical approach utilized in this study also identified key relationships between specific bacterial groups associated with adaptation to warm-season grass pasture and forage nutrients, fecal metabolites, and equine glycemic responses to administration of oral sugar tests.\n",
      "Span[38:39]: \"sugar\" → Chemical (0.8968)\n",
      "********************\n",
      "and Clostridium butyricum in the hindgut microbiome of grazing horses, as these bacteria may play a role in modulation of equine metabolic health.\n",
      "Span[1:3]: \"Clostridium butyricum\" → Species (0.7704)\n",
      "Span[9:10]: \"horses\" → Species (0.5542)\n",
      "********************\n"
     ]
    }
   ],
   "source": [
    "tagger = Classifier.load(\"hunflair\")\n",
    "for key in articleDetails.sections.keys():\n",
    "    sentences = nltk.sent_tokenize(articleDetails.sections[key])\n",
    "    for sentence in sentences:\n",
    "        #print(sentence.replace(\"\\n\",\"\").strip())\n",
    "        sentence1=sentence.replace(\"\\n\",\"\").strip()\n",
    "        sentence = Sentence(sentence1)\n",
    "        tagger.predict(sentence)\n",
    "        if sentence.get_labels()!=[]:\n",
    "            print(sentence1)\n",
    "            for entity in sentence.get_labels():\n",
    "                print(entity)\n",
    "            print(\"*\"*20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "65d3af4b-5478-4e31-8a17-577ca1695615",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(63, 69)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sentence.annotation_layers[\"species\"][1].data_point.start_position,sentence.annotation_layers[\"species\"][1].data_point.end_position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "929466cc-eb8f-4509-acff-2129a331f8d1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Span[1:3]: \"Clostridium butyricum\"'/'Species' (0.7704),\n",
       " 'Span[9:10]: \"horses\"'/'Species' (0.5542)]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sentence.annotation_layers[\"species\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7310728f-17e3-456e-80f1-3c3666fdaddf",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df=pd.DataFrame(columns=[\"sentence\",\"entities\",\"labels\",\"starts\",\"ends\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "32df37f5-46f4-44a5-9b9a-0ecd18c3d717",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "species\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'str' object has no attribute 'data_point'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[22], line 8\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m sent_ann \u001b[38;5;129;01min\u001b[39;00m sentence\u001b[38;5;241m.\u001b[39mannotation_layers:\n\u001b[1;32m      7\u001b[0m     \u001b[38;5;28mprint\u001b[39m(sent_ann)\n\u001b[0;32m----> 8\u001b[0m     en_text\u001b[38;5;241m.\u001b[39mappend(\u001b[43msent_ann\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdata_point\u001b[49m\u001b[38;5;241m.\u001b[39mtext)\n\u001b[1;32m      9\u001b[0m     entity\u001b[38;5;241m.\u001b[39mappend(sent_ann\u001b[38;5;241m.\u001b[39mdata_point\u001b[38;5;241m.\u001b[39mtag)\n\u001b[1;32m     10\u001b[0m     start\u001b[38;5;241m.\u001b[39mappend(sent_ann\u001b[38;5;241m.\u001b[39mdata_point\u001b[38;5;241m.\u001b[39mstart_position)\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'str' object has no attribute 'data_point'"
     ]
    }
   ],
   "source": [
    "\n",
    "st=sent.text\n",
    "en_text=[]\n",
    "entity=[]\n",
    "start=[]\n",
    "end=[]\n",
    "for sent_ann in sentence.annotation_layers.keys():\n",
    "    print(sent_ann)\n",
    "    en_text.append(sent_ann.data_point.text)\n",
    "    entity.append(sent_ann.data_point.tag)\n",
    "    start.append(sent_ann.data_point.start_position)\n",
    "    end.append(sent_ann.data_point.end_position)\n",
    "df.loc[len(df)]=[st,en_text,entity,start,end]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "3fc06f9c-f47b-4c56-8b08-539c94c02260",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sentence</th>\n",
       "      <th>entities</th>\n",
       "      <th>labels</th>\n",
       "      <th>starts</th>\n",
       "      <th>ends</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>and</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Clostridium</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>butyricum</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>in</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>the</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>hindgut</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>microbiome</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>of</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>grazing</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>horses</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>,</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>as</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>these</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>bacteria</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>may</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>play</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>a</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>role</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>in</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>modulation</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>of</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>equine</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>metabolic</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>health</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>.</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       sentence entities labels starts ends\n",
       "0           and       []     []     []   []\n",
       "1   Clostridium       []     []     []   []\n",
       "2     butyricum       []     []     []   []\n",
       "3            in       []     []     []   []\n",
       "4           the       []     []     []   []\n",
       "5       hindgut       []     []     []   []\n",
       "6    microbiome       []     []     []   []\n",
       "7            of       []     []     []   []\n",
       "8       grazing       []     []     []   []\n",
       "9        horses       []     []     []   []\n",
       "10            ,       []     []     []   []\n",
       "11           as       []     []     []   []\n",
       "12        these       []     []     []   []\n",
       "13     bacteria       []     []     []   []\n",
       "14          may       []     []     []   []\n",
       "15         play       []     []     []   []\n",
       "16            a       []     []     []   []\n",
       "17         role       []     []     []   []\n",
       "18           in       []     []     []   []\n",
       "19   modulation       []     []     []   []\n",
       "20           of       []     []     []   []\n",
       "21       equine       []     []     []   []\n",
       "22    metabolic       []     []     []   []\n",
       "23       health       []     []     []   []\n",
       "24            .       []     []     []   []"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45e4d648-32db-4547-9d39-408729182c1c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
